
#Fixes bug with Mac OS update
dyn.load('/Applications/Jalview/jre/Contents/Home/lib/server/libjvm.dylib')

library(Cairo)
library(animation)
this.dir <- dirname(parent.frame(2)$ofile)
this.dir <- paste(c(this.dir,'..'),collapse = '/')
setwd(this.dir)


########
# Colour parameters
########
#
# Master colour scale for heatmap-style plots
my_color_list <- c(
  rgb(1, 0.45, 0.25),
  rgb(0.8, 0.25, 0.25),
  rgb(0, 0, 0),
  rgb(0.25, 0.45, 0.8),
  rgb(0.25, 0.75, 1)
)
# Gene colours for drawing legends
my_gene_colors <- list(
  'PDR5' = rgb(47, 144, 206, maxColorValue = 255),
  'SNQ2' = rgb(179, 231, 172, maxColorValue = 255),
  'YOR1' = rgb(233, 153, 76, maxColorValue = 255),
  'YBT1' = rgb(255, 255, 191, maxColorValue = 255),
  'YCF1' = rgb(166, 7, 13, maxColorValue = 255),
  'BPT1' = rgb(32, 32, 32, maxColorValue = 255)
)
# Three and two gene colour scales generated by my_color_list
blue_black_orange <- grDevices::colorRampPalette(my_color_list)
black_blue <- grDevices::colorRampPalette(my_color_list[3:5])
# Package containing main scripts
#devtools::load_all('../packages/twasAnalysis')
#devtools::document('../packages/twasAnalysis')


########
# Where to read/write files
########
#
# Raw data directory
input_data_directory <- '../data/'
# Processed data directory
output_data_directory <- '../data/output'
# Individual growth experiments are in their own subfolder
# in the input data directory
tecan_output_path <- 'tecan_analysis'
# Genotyping files
all_twas_strains <- 'all_twas_strains.tsv'
extra_genotying_data <- 'extra_genotyping.tsv'
# Map of strain IDs to genotypes
mapping_filename <- 'twas_id_map_fixed.tsv'
# Previously-called genotyping accuracy
genotyping_accuracy_file <- 'leave-one-out-x-validation07192013.tsv'


########
# Run-specific configuration
########
#
# Stores parameters for separate runs
parameter_file <- 'sample_parameters.tsv'
# Sample name
sample <- 'X14.Nov'
#
# Parameters from this file are set below
#
# Read sample-specific variables
setwd(input_data_directory)
sample_pars <- read.table(parameter_file, stringsAsFactors = F)
sample_vals <- sample_pars[, sample]
# Set sample-specific variables from parameter file
# Sub-folder to create 'deliverables' (figures, data, etc)
output_results_directory <- sample_vals[1]
# What to call the run-specific resistances generated (deprecated)
resistance_output_prefix <- sample_vals[2]
# What to call the run-specific linear model
lm_output_file <- sample_vals[3]
# Processed sequencing data to read
sequencing_filename <- sample_vals[4]
# What to call the run-specific resistances generated (MATa)
A_resistance_filename <- sample_vals[5]
# What to call the run-specific resistances generated (MATalpha)
alpha_resistance_filename <- sample_vals[6]
# Name of file specifying what the solvent control is for each drug
# if 'none', default parameters are used for many functions
drug_control_pair <- sample_vals[7]
if (drug_control_pair == 'none') {
  drug_control_pair <- NULL
}

########
# Pre-run maintenance
########
#
# global control of what metric is used for analysis
#  'resistance' - compares growth in each drug to control
#  'growth' - takes growth in each drug as-is
growth_metric <- 'growth'
# Create Sub directories for output if they don't exist
dir.create(output_data_directory, showWarnings = FALSE)
dir.create(output_results_directory, showWarnings = FALSE)
# Linear model input results, defaults to the same file as the output file
lm_input_file <- lm_output_file


########



setwd(this.dir)
setwd(input_data_directory)
input_file <- read.table(sequencing_filename,
                         head = T,
                         row.names = 1)
mapping_file <- read.table(mapping_filename,
                           head = T,
                           row.names = 1)
drug_control_list <- NULL

#Creates a list of what control to use for each drug, if it is provided
#Otherwise it defaults to 'NULL'
if (!is.null(drug_control_pair)) {
  drug_control_list <- list()
  drug_control_pair_file <- read.table(file = drug_control_pair)
  for (i in 1:nrow(drug_control_pair_file)) {
    drug <- as.vector(drug_control_pair_file[i, 1])
    ctrl <- as.vector(drug_control_pair_file[i, 2])
    drug_control_list[[drug]] <- ctrl
  }
}

rhombus_integration <- function(x,y){
  sum <- 0
  for(i in 2:length(x)){
    base <- x[i] - x[i - 1]
    mid_height <- mean(c(y[i],y[i-1]))
    sum <- sum + base*mid_height
  }
  return(sum)
}




create_frequency_list <- function(input_file,
                                  time_avail_list,
                                  drugs=NULL,
                                  A_indeces,
                                  alpha_indeces,
                                  mating_types=c('A','alpha')){
  made_ctrl <- list()
  made_ctrl[['alpha']] <- F
  made_ctrl[['A']] <- F
  
  frequency_list <- list()
  frequency_list[['depth']] <- list()
  #count_list <- list()
  #modality_matrix <- list()
  
  if(is.null(drugs)){
    drugs <- names(time_avail_list)
  }
  #Record appropriate values for each sample
  for(drug in drugs){
    for(strain_ind in mating_types){
      time_points <- time_avail_list[[drug]][[strain_ind]]
      for(time_point in time_points){
        #
        
        cond_name <- paste(c(drug,strain_ind,time_point),collapse='_')
        condA <- paste(c('A',time_point,'_',drug),collapse='')
        condalpha <- paste(c('alpha',time_point,'_',drug),collapse='')
        
        
        if(strain_ind == 'A'){
          As <- A_indeces
          alphas <- alpha_indeces[1]
        }
        if(strain_ind == 'alpha'){
          As <- A_indeces[1]
          alphas <- alpha_indeces
        }      
        if(strain_ind == 'both'){
          As <- A_indeces
          alphas <- alpha_indeces
        }      
        
        
        output_file_A <- input_file[As,,drop=F]
        output_file_alpha <- input_file[alphas,,drop=F]
        
        merged_file_A <- merge(mapping_file,output_file_A,by="row.names",all.x=F,all.y=F)
        merged_file_alpha <- merge(mapping_file,output_file_alpha,by="row.names",all.x=F,all.y=F)
        
        
        response_index <- grep(condA,colnames(merged_file_A),val=T)
        response_index_alpha <- grep(condalpha,colnames(merged_file_alpha),val=T)
        
        print('')
        print('Response indeces')
        print(response_index)
        print(response_index_alpha)
        print('END')
        print('')
        
        response_index <- grep(condA,colnames(merged_file_A))
        response_index_alpha <- grep(condalpha,colnames(merged_file_alpha))
        
        ctrlA <- grep(paste(c('T0_A'),collapse=''),colnames(merged_file_A))
        ctrlalpha <- grep(paste(c('T0_alpha'),collapse=''),colnames(merged_file_alpha))        
        
        #Counts for current condition
        #v1 is current condition
        #v2 is control
        
        v1 <- apply(merged_file_A[,response_index,drop=F],1,mean)
        v2 <- apply(merged_file_A[,ctrlA,drop=F],1,mean)
        
        v1_alpha <- apply(merged_file_alpha[,response_index_alpha,drop=F],1,mean)
        v2_alpha <- apply(merged_file_alpha[,ctrlalpha,drop=F],1,mean)
        
        #Strain must be present in control condition to process
        merged_file_A <- merged_file_A[(v2 > 0),]
        merged_file_alpha <- merged_file_alpha[(v2_alpha > 0),]
        
        
        
        if(strain_ind == 'A'){
          #Add pseudocounts as well
          v1_n <- apply(merged_file_A[,response_index,drop=F],1,mean) + 0.5
          v2_n <- apply(merged_file_A[,ctrlA,drop=F],1,mean) + 0.5
          
          sum1 <- sum(v1_n)
          sum2 <- sum(v2_n)
        }
        if(strain_ind == 'alpha'){
          #Repeat of loop for 'A', see comments there
          v1_n_alpha <- apply(merged_file_alpha[,response_index_alpha,drop=F],1,mean) + 0.5
          v2_n_alpha <- apply(merged_file_alpha[,ctrlalpha,drop=F],1,mean) + 0.5
          
          sum1_alpha <- sum(v1_n_alpha)
          sum2_alpha <- sum(v2_n_alpha)
        }
        
        
        if(strain_ind == 'A' & made_ctrl[['A']] == F){
          
          made_ctrl[['A']] == T
          simple_resistance <- v2_n/sum2
          names(simple_resistance) <- merged_file_A[,1]
          counts <- v2_n
          names(counts) <- merged_file_A[,1]
          
          frequency_list[['depth']][['CTRL']][['A']] <- c(frequency_list[['depth']][['CTRL']][['A']],list())
          frequency_list[['depth']][['CTRL']][['A']][['0']] <- sum2
          frequency_list[['CTRL']][['A']] <- simple_resistance
          #count_list[['CTRL']][['A']] <- counts
          
        }
        
        if(strain_ind == 'alpha' & made_ctrl[['alpha']] == F){
          
          made_ctrl[['alpha']] == T
          simple_resistance <- v2_n_alpha/sum2_alpha
          names(simple_resistance) <- merged_file_alpha[,1]
          counts <- v2_n_alpha
          names(counts) <- merged_file_alpha[,1]
          
          
          frequency_list[['depth']][['CTRL']][['alpha']] <- c(frequency_list[['depth']][['CTRL']][['alpha']],list())
          frequency_list[['depth']][['CTRL']][['alpha']][['0']] <- sum2_alpha
          frequency_list[['CTRL']][['alpha']] <- simple_resistance
          #count_list[['CTRL']][['alpha']] <- counts
          
          
        }
        
        if(strain_ind == 'alpha'){
          simple_resistance <- v1_n_alpha/sum1_alpha
          counts <- v1_n_alpha
          depth <- sum1_alpha
        }      
        
        if(strain_ind == 'A'){
          simple_resistance <- v1_n/sum1
          counts <- v1_n
          depth <- sum1
        }    
        
        
        frequency_list[["depth"]] <- c(frequency_list[["depth"]], list())
        frequency_list[["depth"]][[drug]] <- c(frequency_list[["depth"]][[drug]], list())
        frequency_list[["depth"]][[drug]][[strain_ind]] <- c(frequency_list[["depth"]][[drug]][[strain_ind]], list())
        
        frequency_list[["depth"]][[drug]][[strain_ind]][[time_point]] <- depth
        
        
        #print(frequency_list[["depth"]][[drug]][[strain_ind]][[time_point]])
        
        frequency_list[[drug]][[time_point]][[strain_ind]] <- simple_resistance
        #count_list[[drug]][[time_point]][[strain_ind]] <- counts
        
      }
    }
  }
  return(frequency_list)
}



make_preliminary_timepoint_availability_list <- function(input_file,drugs=NULL,mating_types=c('A','alpha')){
  time_avail_list <- list()
  if(is.null(drugs)){
    drugs <- get_all_drugs(input_file)
  }
  for(drug in drugs){
    time_avail_list[[drug]] <- list()
    drug_matching_names <- colnames(input_file)[grep(drug,colnames(input_file))]
    drug_matching_names <- sapply(drug_matching_names,function(x){substr(x,start=1,stop=8)})
    for(mating_type in mating_types){
      mating_matching_names <- drug_matching_names[grep(mating_type,drug_matching_names)]
      time_points <- sapply(mating_matching_names,function(name){
        mating_time <- strsplit(name,split='_')[[1]][1]
        time <- strsplit(mating_time,mating_type)[[1]][2]
        return(time)
      })
      
      #eyo <<- time_points
      #time_points <- intersect(time_points,c('0','5','10'))
      #eyo2 <<- time_points#,'10'))#,'10','15'))
      time_avail_list[[drug]][[mating_type]] <- sapply(unique(sort(as.numeric(time_points))),toString)
    }
  }
  return(time_avail_list)
}

make_timepoint_availability_list <- function(input_file,drugs=NULL,drug_control_list=NULL){
  
  preliminary_availability <- make_preliminary_timepoint_availability_list(input_file,drugs)
  return(preliminary_availability)
  
  #pruned_availability <- merge_timepoint_availability_list(preliminary_availability,drug_control_list)
  
  
  return(pruned_availability)
}



make_growth_list <- function(input_file,mapping_file,drugs=NULL,growth_metric='resistance',drug_control_list=NULL){
  #mating_types <- c('A','alpha')
  alpha_indeces <- grep('alpha',sapply(rownames(input_file),function(strain){mapping_file[strain,'Plate']}))
  A_indeces <- grep('alpha',sapply(rownames(input_file),function(strain){mapping_file[strain,'Plate']}),invert=T)
  
  if(is.null(drugs)){
    if(is.null(drug_control_list)){
      drugs <- get_all_drugs(input_file)
    }
    else{
      drugs <- names(drug_control_list)
      drugs <- c(drugs,unique(unlist(drug_control_list)))
    }
  }
  
  
  print(drugs)
  time_avail_list <- make_timepoint_availability_list(input_file,drugs,drug_control_list)
  
  frequency_list <- create_frequency_list(input_file,
                                           time_avail_list,
                                           drugs,
                                           A_indeces,
                                           alpha_indeces)
  #return(frequency_list)
  growth_list <- create_resistance_list(frequency_list,time_avail_list,metric=growth_metric,drugs=drugs,drug_control_list=drug_control_list)
  return(growth_list)
}



create_resistance_list <- function(frequency_list,
                                   time_avail_list,
                                   metric='resistance',
                                   drugs,
                                   mating_types=c('A','alpha'),
                                   drug_control_list=NULL){
  if(is.null(drug_control_list)){
    drug_control_list <- list()
    for(drug in drugs){
      drug_control_list[[drug]] <- 'CTRL'
    }
  }
  controls <- c(unique(unlist(drug_control_list)),'CTRL','depth')
  drugs <- drugs[!(drugs %in% controls)]
  
  if(metric == 'resistance'){
    drugs <- drugs
  }else if (metric == 'growth'){
    drugs <- unique(c(drugs, unlist(drug_control_list)))
  }
  
  resistance_list <- list()
  for(mating_type in mating_types){
    for(drug in drugs){
      strain_names <- names(frequency_list[['CTRL']][[mating_type]])
      resistance <- fitness_estimator(drug,mating_type,frequency_list,time_avail_list,metric,ctrl=drug_control_list[[drug]])
      #if(na_fix == T){
      #  resistance[!is.finite(resistance)] <- min(resistance[is.finite(resistance)])
      #}
      names(resistance) <- strain_names
      resistance_list[[drug]][[mating_type]] <- resistance[!is.na(resistance)]
      
    }
  }
  return(resistance_list)
}

fitness_estimator <- function(drug,mating_type,frequency_list,time_avail_list,metric='resistance',ctrl='CTRL',healthy_cutoff=0.5){
  frequency_list <- frequency_list
  print(drug)
  print(mating_type)
  indeces <- length(frequency_list[[ctrl]][[mating_type]])
  time_points <- as.numeric(time_avail_list[[drug]][[mating_type]])
  fitness <- sapply(1:indeces,function(i){
    
    
    
    t0 <- frequency_list[[ctrl]][[mating_type]][[i]]
    d0 <- frequency_list[['depth']][['CTRL']][[mating_type]][[1]]
    #Calculate a curve for the drug
    #Then model a growth rate (k) by assuming
    #Observed curve is obtained by function C_0*e^(kt)
    #print('Here')
    
    freqs <- c(t0)
    depths <- c(d0)
    for(time_point in time_points){
      tn <- frequency_list[[drug]][[toString(time_point)]][[mating_type]][[i]]
      dn <- frequency_list[['depth']][[drug]][[mating_type]][[toString(time_point)]]
      freqs <- c(freqs,tn)
      depths <- c(depths,dn)
    }
    
    #print(depths)
    #depths <- c()
    
    freq_list_global <<- frequency_list
    freqs_global <<- freqs
    depths_global <<- depths
    ey <<- 5
    
    # g <- optim_growth(freqs,depths,time_points)
    # 
    # stop()  
    # if(metric == 'growth'){
    #   return(g)
    # }
    # else{
    #   freqs <- c(t0)
    #   depths <- c(d0)
    #   for(time_point in time_points){
    #     tn <- frequency_list[[ctrl]][[toString(time_point)]][[mating_type]][[i]]
    #     dn <- frequency_list[['depth']][[ctrl]][[mating_type]][[toString(time_point)]]
    #     freqs <- c(freqs,tn)
    #     depths <- c(depths,dn)
    #   }
    #   
    #   g2 <- optim_growth(freqs,depths,time_points)^1.91
    #   
    #   
    #   #if(g2 < healthy_cutoff){
    #   #  return(0)
    #   #}
    #   return((g+0.1)/(g2+0.1))
    # }
    # 
    #freqs_ctrl <- c(t0)
    ##depths <- c(d0)
    #for(time_point in time_points){
    #  tn <- frequency_list[[ctrl]][[toString(time_point)]][[mating_type]][[i]]
    #  dn <- frequency_list[['depth']][[ctrl]][[toString(time_point)]][[mating_type]]
    #  freqs <- c(freqs,tn)
    #  depths <- c(depths,dn)
    #}
    
    
    #return(g)
    
    #stop()
    
    estims_drug <- c(t0)
    for(time_point in time_points){
      tn <- frequency_list[[drug]][[toString(time_point)]][[mating_type]][[i]]
      # tn_ctrl <- frequency_list[[ctrl]][[toString(time_point)]][[mating_type]][[i]]
      
      estims_drug <- c(estims_drug,(2^time_point)*(tn))
    }
    #Try
    estims_drug <- estims_drug
    
    times <- c(time_points)
    
    time_points_global <<- time_points
    
    gen_to_time <- list('5' = 1,
                        '10' = 1.5,
                        '15' = 2,
                        '20' = 2.5)
    
    
    
    
    time_point_char <- as.character(time_points)
    
    real_times <- gen_to_time[time_point_char]
    real_times <- as.vector(unlist(real_times))
    
    real_times <- time_points
    
    
    
    auc_drug <- rhombus_integration(c(0,real_times),estims_drug)
    
    
    
    #auc_drug <- (rhombus_integration(c(0,time_points),estims_drug))
    
   # stop()
    #auc_drug <- rhombus_integration(time_points,estims_drug)
    
    
 #   stop()
    
    # exp_growth_rate_drug <- try_solve_auc(auc_drug,t0,max(time_points))#log(auc_drug/t0)/max(time_points)
    
    
    #print('evaluating drug')
    #print(estims_drug)
    #estims_drug[estims_drug == 0] <- min(estims_drug[estims_drug != 0])
    #Rule-of-thumb to not allow negative growth
    #if(estims_drug[length(estims_drug)] < estims_drug[1]){
    #  exp_growth_rate_drug <- 0
    #}else{
    #  print('trying')
    #  exp_growth_rate_drug <- max(glm(estims_drug~times,family=gaussian(link='log'))$coefficients[2],0)
    #}
    if(metric=='resistance'){
      estims_ctrl <- c(t0)
      for(time_point in time_points){
        tn <- frequency_list[[ctrl]][[toString(time_point)]][[mating_type]][[i]]
        estims_ctrl <- c(estims_ctrl,(2^time_point)*(tn))
      }
      #exp_growth_rate_drug <- max(glm(estims_drug~times,family=gaussian(link='log'),start=c(0,0))$coefficients[2],0)
      
      #time_
      auc_ctrl <- rhombus_integration(c(0,real_times),estims_ctrl)
      #auc_ctrl <- (rhombus_integration(c(0,time_points),estims_ctrl))
      
      
      # exp_growth_rate_ctrl <- try_solve_auc(auc_ctrl,t0,max(time_points))
      #exp_growth_rate_ctrl <- log(auc_ctrl/t0)/max(time_points)
      #print('evaluating_ctrl')
      #print(estims_ctrl)
      
      #Again do not allow negative growth
      #if(estims_ctrl[length(estims_ctrl)] < estims_ctrl[1]){
      #  exp_growth_rate_ctrl <- 0
      #}else{
      #  print('trying')
      #  exp_growth_rate_ctrl <- max(glm(estims_ctrl~times,family=gaussian(link='log'))$coefficients[2],0)
      #}
      #estims_ctrl[estims_ctrl == 0] <- min(estims_ctrl[estims_ctrl != 0])
      #exp_growth_rate_ctrl <- max(glm(estims_ctrl~times,family=gaussian(link='log'))$coefficients[2],0)
    }
    if(metric=='resistance'){
      #metr <- log10(auc_drug/auc_ctrl)#/max(time_points)
      #return(metr)
      
      #print('drug')
      
      #return(exp_growth_rate_drug/exp_growth_rate_ctrl)
      #print('ctrl')
      #print(exp_growth_rate_ctrl)
      #print('ratio')
      #if(exp_growth_rate_ctrl > healthy_cutoff){
      #metr <- exp_growth_rate_drug/exp_growth_rate_ctrl #/max(time_points)
      #metr <- log2(auc_drug/auc_ctrl)/max(time_points)
      #  #  ratio <- exp_growth_rate_drug/exp_growth_rate_ctrl
      #  #Small number so taking log doesn't crash
      #  ratio <- max(ratio,1e-5)
      #}
      #else{
      #  metr <- NA
      #}
      #print(ratio)
      #if(identical(metr,0)){
      #  metr <- 1e-4
      #}
      #k1 <- log2((auc_drug*log(2)+t0)/t0)/max(time_points)
      #k2 <- log2((auc_ctrl*log(2)+t0)/t0)/max(time_points)
      
      #metr <- auc_drug
      #metr <- log2((auc_drug*log(2)+1))/max(time_points)
      
      
      metr <- log2(auc_drug/auc_ctrl)/max(time_points)
      
      return(metr)
      #if(max(exp_growth_rate_drug,exp_growth_rate_ctrl) > 0){
      #  print(exp_growth_rate_drug/exp_growth_rate_ctrl)
      #  return(exp_growth_rate_drug/exp_growth_rate_ctrl)
      #}
      #else{
      #  print('Undef')
      #  return(exp_growth_rate_drug/exp_growth_rate_ctrl)
      #}
      
    }
    if(metric=='growth'){
      #metr <- log10(estims_drug[2]/estims_drug[1])
      metr <- auc_drug#log10(auc_drug)/max(time_points)
      
      return(metr)
      
      #return(log10(auc_drug)/max(time_points))
      #return(log10(auc_drug))#/auc_ctrl))
      #return(exp_growth_rate_drug)
    }
  })
  return(fitness)
}




optim_growth <- function(freqs,depths,generations,interval_vec=c(0,2)){
  growth_probability <- function(g,init_freq,freq,ngen,depth){
    expected_freq <- init_freq*2^(g*ngen - ngen)
    if(expected_freq >= 1){
      expected_freq <- 1
    }
    ll <- -dbinom(round(freq*depth),depth,expected_freq,log=T)
    return(ll)
  }
  
  init_freq <- freqs[1]
  derived_freqs <- freqs[2:length(freqs)]
  
  joint_probability <- function(g){
    return(sum(sapply(1:length(derived_freqs),function(i){
      freq <- derived_freqs[i]
      ngen <- generations[i]
      depth <- depths[i]
      return(growth_probability(g,init_freq,freq,ngen,depth))
    })))
  }
  
  g_vec <- seq(interval_vec[1],interval_vec[2],length.out = 10)
  
  probs <- sapply(g_vec,function(g){joint_probability(g)})
  
  initial_answer <- g_vec[which.min(probs)]
  
  interval <- g_vec[2] - g_vec[1]
  
  #Refine the interval
  refined_answer <- tryCatch(optimize(joint_probability,
                                      interval = c(max(c(initial_answer - interval,interval_vec[1])),
                                                   min(c(initial_answer + interval,interval_vec[2]))))$min,
                             warning = function(x){initial_answer})
  
  return(refined_answer)
  
  g_vec <- seq(interval_vec[1],interval_vec[2],length.out = 100)
  
  probs <- sapply(g_vec,function(g){joint_probability(g)})
  
  initial_answer <- which.min(probs)
  
  #print(probs)
  #print(initial_answer)
  
  #Refine the interval
  if(g_vec[initial_answer] > interval_vec[1] &
     g_vec[initial_answer] < interval_vec[2]) {
    informed_lower <- g_vec[initial_answer - 1]
    informed_upper <- g_vec[initial_answer + 1]
    refined_answer <- optimize(joint_probability,
                               interval = c(informed_lower,informed_upper))$min
    return(refined_answer)
  }
  else{
    return(g_vec[initial_answer])
  }
}


input_file <- input_file[,grep('DN',colnames(input_file),val=T)]

growth_list <- make_growth_list(
  input_file,
  mapping_file,
  growth_metric = growth_metric,
  drugs = c('CTRL',
            'benomyl',
            'fluconazole',
            'ketoconazole',
            'miconazole',
            'itraconazole',
            'beauvericin',
            'colchicine',
            'cycloheximide',
            'mitoxantrone',
            'camptothecin',
            'tamoxifen',
            'bisantrene'),
  drug_control_list = drug_control_list
)

##recapitulate javier data


drug <- 'fluconazole'
mating_type <- 'alpha'

strain_names <- names(freq_list_global$CTRL[[mating_type]])
wts <- which(apply(mapping_file[strain_names,2:17],1,sum) == 0)


depths <- freq_list_global$depth$CTRL[[mating_type]]
count_t0 <- depths$`0`
count_t0 <- freq_list_global$CTRL[[mating_type]]*count_t0

depths <- freq_list_global$depth[[drug]][[mating_type]]

count_t5 <- depths$`5`
count_t5 <- freq_list_global[[drug]]$`5`[[mating_type]]*count_t5

count_t10 <- depths$`10`
count_t10 <- freq_list_global[[drug]]$`10`[[mating_type]]*count_t10

count_t15 <- depths$`15`
count_t15 <- freq_list_global[[drug]]$`15`[[mating_type]]*count_t15

count_t20 <- depths$`20`
count_t20 <- freq_list_global[[drug]]$`20`[[mating_type]]*count_t20

good_strains_t0 <- count_t0 > 100
good_strains_t5 <- count_t5 > 100
good_strains_t10 <- count_t10 > 100
good_strains_t15 <- count_t15 > 100
#good_strains_t20 <- count_t20 > 50

all_good <- good_strains_t0 & good_strains_t5 & good_strains_t10 & good_strains_t15# & good_strains_t20

freq_t0 <- freq_list_global$CTRL[[mating_type]]#[all_good]
freq_t5 <- freq_list_global[[drug]]$`5`[[mating_type]]#[all_good]

freq_t10 <- freq_list_global[[drug]]$`10`[[mating_type]]#[all_good]
freq_t15 <- freq_list_global[[drug]]$`15`[[mating_type]]#[all_good]
freq_t20 <- freq_list_global[[drug]]$`20`[[mating_type]]#[all_good]



consolidated_list <- list(
  'counts'=list('0'=count_t0,
                '5'=count_t5,
                '10'=count_t10,
                '15'=count_t15,
                '20'=count_t20),
  'frequencies'=list(
    '0'=freq_t0,
    '5'=freq_t5,
    '10'=freq_t10,
    '15'=freq_t15,
    '20'=freq_t20
  )
)

var_vec  <- c()
g_matrix <- c()
ratio_matrix <- c()
time_points <- c('0','5','10','15')#,'20')



for(time_point1 in time_points){
  for(time_point2 in time_points){
    
    t1 <- as.numeric(time_point1)
    t2 <- as.numeric(time_point2)
    if(t2 > t1){
      print(c(t2,t1))
      time_diff <- t2 - t1
      comparison_name <- paste(c(time_point2, time_point1),collapse='_')
      count_t1 <- consolidated_list[['counts']][[time_point1]]
      count_t2 <- consolidated_list[['counts']][[time_point2]]
      
      freq_t1 <- consolidated_list[['frequencies']][[time_point1]]
      freq_t2 <- consolidated_list[['frequencies']][[time_point2]]
      
      goodness_criteria <- (count_t1 > (2*(2^time_diff) + 1)) | (count_t2 > 50 & count_t1 > 50)
      
      
      #good_wts <- wts[wts %in% which(goodness_criteria)]
      
      ratio <- log2(freq_t2/freq_t1)
      ratio[!goodness_criteria] <- NA
      
      ratio_matrix <- cbind(ratio_matrix, ratio)
      #stop()
      #print(comparison_name)
      #print(quantile(ratio[goodness_criteria],probs=c(0.0),na.rm=T))
      #print(time_diff)
      dubs <- ratio + time_diff
      
      dubs_good_strains <- dubs[all_good]
      
      dubs <- dubs/median(dubs_good_strains)
      
      #dubs <- (ratio + time_diff)/(median(ratio[good_wts]) + time_diff)
      
      #print(median(dubs[which(goodness_criteria[wts])]))
      
      #print(var(dubs[good_wts]))
      
      #var_vec <- c(var_vec,var(dubs[good_wts]))
      
      dubs[!goodness_criteria] <- NA
      
     # print(comparison_name)
    #  print(var(dubs[which(goodness_criteria[wts])],na.rm=T))
      
      ncol1 <- ncol(g_matrix)
      
      g_matrix <- cbind(g_matrix,dubs)
      
      ncol2 <- ncol(g_matrix)

      if(identical(ncol1,ncol2)){
        stop()
      }
            
      print(dim(g_matrix))
      
      colnames(g_matrix)[ncol(g_matrix)] <- comparison_name
      
      colnames(ratio_matrix)[ncol(ratio_matrix)] <- comparison_name
      
      
      #print(sum(goodness_criteria))
      
    }
  }
}

stop()

median_growth <- apply(g_matrix,1,function(x){
  mean(x,na.rm=T)#weighted.mean(x,w=var_vec,na.rm=T)
})


mapping_file_matched <- mapping_file[strain_names,]


hist(median_growth,breaks=100)

stop()

g_0_5 <- log2(freq_t5/freq_t0)# + 3
g_5_10 <- log2(freq_t10/freq_t5)# + 5
g_10_15 <- log2(freq_t15/freq_t10)# + 5
g_15_20 <- log2(freq_t20/freq_t15)# + 5

g_0_5[!good_strains_t0] <- NA
g_5_10[!good_strains_t5] <- NA
g_10_15[!good_strains_t10] <- NA
g_15_20[!good_strains_t15] <- NA



w_0_5 <- (g_0_5 + 5)/(median(g_0_5[all_good]) + 5)
w_5_10 <- (g_5_10 + 5)/(median(g_5_10[all_good]) + 5)
w_10_15 <- (g_10_15 + 5)/(median(g_10_15[all_good]) + 5)
w_15_20 <- (g_15_20 + 5)/(median(g_15_20[all_good]) + 5)



w_all <- cbind(w_0_5,w_5_10,w_10_15,w_15_20)

w_averaged <- apply(w_all,1,mean,na.rm=T)

stop()





gees <- cbind(g_0_5,g_5_10,g_10_15,g_15_20)

#normalization_factor <- apply(gees[all_good,],2,median)
#var_normalization_factor <- apply(gees[all_good,],2,sd)
#var_normalization_factor <- var_normalization_factor/min(var_normalization_factor)




gees_normalized_guess <- sapply(1:ncol(gees),function(i){
  wt_diff <- median(gees[all_good,i])
  new_growth <- (gees[,i] + 5)/(5 + wt_diff)
  return(new_growth)
})
var_normalization_factor <- apply(gees_normalized_guess[all_good,],2,sd)


gees_normalized <- sapply(1:ncol(gees),function(i){
  
  
  wt_diff <- median(gees[all_good,i])
  
  #Derived that log2(r_wt) = t_wt - t_pool
  #t_pool is assigned as 5
  
  #Rescale generations so that variance = minimum variance amongst timepoints
  opt_gen <- optimize(function(x){
    #new_growth <- (gees[,i] + 5*var_normalization_factor[i])/(5*var_normalization_factor[i] + wt_diff)
    
    new_growth <- (gees[,i] + 5*x)/(5*x + wt_diff)
    
    return(abs(sd(new_growth[all_good]) - median(var_normalization_factor)))
    
  },interval=c(0.8,2))
  
  print(opt_gen$min)
  #print(opt_gen$obj)
  opt_gen <- opt_gen$min
  #opt_gen <- 1
  
  
  new_growth <- (gees[,i] + 5*opt_gen)/(5*opt_gen + wt_diff)
  
  #print(sd(new_growth[all_good]))
  
  return(new_growth)
  
  
  #(gees[,i] - normalization_factor[i] + 1)
})



averaged_growth <- apply(gees_normalized,1,mean,na.rm=T)