
#Fixes bug with Mac OS update
dyn.load('/Applications/Jalview/jre/Contents/Home/lib/server/libjvm.dylib')

library(Cairo)
library(animation)
this.dir <- dirname(parent.frame(2)$ofile)
this.dir <- paste(c(this.dir,'..'),collapse = '/')
setwd(this.dir)


########
# Colour parameters
########
#
# Master colour scale for heatmap-style plots
my_color_list <- c(
  rgb(1, 0.45, 0.25),
  rgb(0.8, 0.25, 0.25),
  rgb(0, 0, 0),
  rgb(0.25, 0.45, 0.8),
  rgb(0.25, 0.75, 1)
)
# Gene colours for drawing legends
my_gene_colors <- list(
  'PDR5' = rgb(47, 144, 206, maxColorValue = 255),
  'SNQ2' = rgb(179, 231, 172, maxColorValue = 255),
  'YOR1' = rgb(233, 153, 76, maxColorValue = 255),
  'YBT1' = rgb(255, 255, 191, maxColorValue = 255),
  'YCF1' = rgb(166, 7, 13, maxColorValue = 255),
  'BPT1' = rgb(32, 32, 32, maxColorValue = 255)
)
# Three and two gene colour scales generated by my_color_list
blue_black_orange <- grDevices::colorRampPalette(my_color_list)
black_blue <- grDevices::colorRampPalette(my_color_list[3:5])
# Package containing main scripts
#devtools::load_all('../packages/twasAnalysis')
#devtools::document('../packages/twasAnalysis')


########
# Where to read/write files
########
#
# Raw data directory
input_data_directory <- '../data/'
# Processed data directory
output_data_directory <- '../data/output'
# Individual growth experiments are in their own subfolder
# in the input data directory
tecan_output_path <- 'tecan_analysis'
# Genotyping files
all_twas_strains <- 'all_twas_strains.tsv'
extra_genotying_data <- 'extra_genotyping.tsv'
# Map of strain IDs to genotypes
mapping_filename <- 'twas_id_map_fixed.tsv'
# Previously-called genotyping accuracy
genotyping_accuracy_file <- 'leave-one-out-x-validation07192013.tsv'


########
# Run-specific configuration
########
#
# Stores parameters for separate runs
parameter_file <- 'sample_parameters.tsv'
# Sample name
sample <- 'X14.Nov'
#
# Parameters from this file are set below
#
# Read sample-specific variables
setwd(input_data_directory)
sample_pars <- read.table(parameter_file, stringsAsFactors = F)
sample_vals <- sample_pars[, sample]
# Set sample-specific variables from parameter file
# Sub-folder to create 'deliverables' (figures, data, etc)
output_results_directory <- sample_vals[1]
# What to call the run-specific resistances generated (deprecated)
resistance_output_prefix <- sample_vals[2]
# What to call the run-specific linear model
lm_output_file <- sample_vals[3]
# Processed sequencing data to read
sequencing_filename <- sample_vals[4]
# What to call the run-specific resistances generated (MATa)
A_resistance_filename <- sample_vals[5]
# What to call the run-specific resistances generated (MATalpha)
alpha_resistance_filename <- sample_vals[6]
# Name of file specifying what the solvent control is for each drug
# if 'none', default parameters are used for many functions
drug_control_pair <- sample_vals[7]
if (drug_control_pair == 'none') {
  drug_control_pair <- NULL
}

########
# Pre-run maintenance
########
#
# global control of what metric is used for analysis
#  'resistance' - compares growth in each drug to control
#  'growth' - takes growth in each drug as-is
growth_metric <- 'growth'
# Create Sub directories for output if they don't exist
dir.create(output_data_directory, showWarnings = FALSE)
dir.create(output_results_directory, showWarnings = FALSE)
# Linear model input results, defaults to the same file as the output file
lm_input_file <- lm_output_file


########



setwd(this.dir)
setwd(input_data_directory)
input_file <- read.table(sequencing_filename,
                         head = T,
                         row.names = 1)
mapping_file <- read.table(mapping_filename,
                           head = T,
                           row.names = 1)
drug_control_list <- NULL


drugs <- unique(sapply(colnames(input_file),function(name){
  strsplit(name,split='_')[[1]][2]
}))
drugs <- unique(drugs)
drugs <- drugs[!(grepl('CTRL',drugs))]
drugs <- drugs[!(drugs %in% c('A','alpha'))]

drugs <- c('cycloheximide')#,'ketoconazole')##'itraconazole',,'ketoconazole')
mating_types <- c('A','alpha')
tags <- c('UP','DN')
time_points <- c('0','5','10','15','20')

estimate_log2_ratio_variance <- function(count2,count1){
  ((sqrt(count2)/count2)^2 + (sqrt(count1)/count1)^2)/(count2/count1 + log(2))
}


calculate_normalized_dubs <- function(counts,
                                      pool_gens,
                                      variance_scaling = T){
  print(pool_gens)
  
  all_good <- apply(counts, 1, min) > 10
  g_matrix <- c()
  var_matrix <- c()
  
  freqs <- apply(counts,2,function(x){x/sum(x)})
  
  for (i in 1:(length(pool_gens) - 1)){
    print(i)
    time_point1 <- pool_gens[i]
    for (j in i + 1){#1:length(pool_gens)){ #i + 1
      time_point2 <- pool_gens[j]
      t1 <- as.numeric(time_point1)
      t2 <- as.numeric(time_point2)
      if (t2 > t1) {
        time_diff <- t2 - t1
        #time_diff_wt <- wt_gens[j] - wt_gens[i]
        
        
        
        
        count_t1 <- counts[, i]
        count_t2 <- counts[, j]
        
        freq_t1 <- freqs[, i]
        freq_t2 <- freqs[, j]
        
        goodness_criteria <-
          (count_t1 > (2*(2 ^ time_diff))) |
          (count_t2 > 30 & count_t1 > 30)
        
        ratio <- log2(freq_t2 / freq_t1)
        ratio[!goodness_criteria] <- NA
        
        time_diff <- max(c(time_diff,-quantile(ratio, probs=0.1, na.rm=T)))
        
        #stop()
        # print(time_diff)
        # print(time_diff_wt)
        
        variance <- estimate_log2_ratio_variance(count_t2,count_t1)
        var_matrix <- cbind(var_matrix,variance/time_diff)
        
        
        growth <- (ratio + time_diff)# / time_diff_wt
        
        g_matrix <- cbind(g_matrix, growth)
        
        comparison_name <-
          paste(c(i, j), collapse = '_')
        colnames(g_matrix)[ncol(g_matrix)] <- comparison_name
      }
    }
  }
  
  #g_matrix_global <<- g_matrix
  
  
  
  norm_factor <- apply(g_matrix,2,function(x){
    median(x[all_good],na.rm=T)
  })
  
  #print(norm_factor)
  
  norm_factor_kept <- sapply(1:(length(pool_gens) - 1),function(i){
    name <- paste(c(i, i + 1), collapse = '_')
    norm_factor[name]
  })
  
  #stop()
  
  g_matrix <- apply(g_matrix,2,function(x){
    x/median(x[all_good],na.rm=T)
  })
  
  #print(head(g_matrix))
  
  
  #median_g <- apply(g_matrix,1,function(x){ median(x,na.rm=T)})
  
  median_g <- sapply(1:nrow(g_matrix),function(i){
    if(variance_scaling == T){
      crit <- !is.na(g_matrix[i,]) &  !is.na(var_matrix[i,])
      
      
      return(weighted.mean(g_matrix[i,crit],w=1/(var_matrix[i,crit]),na.rm=T))
    }
    return(mean(g_matrix[i, ], na.rm = T))
  })
  
  #median_g <- median_g/median(median_g[all_good])
  
  median_g <- sapply(1:nrow(g_matrix),function(i){
    if(variance_scaling == T){
      crit <- !is.na(g_matrix[i,]) &  !is.na(var_matrix[i,])
      
      
      return(weighted.mean(g_matrix[i,crit],w=1/(var_matrix[i,crit]),na.rm=T))
    }
    return(mean(g_matrix[i, ], na.rm = T))
  })
  
  #Quick correction
  median_g[median_g < 0] <- 0
  #median_g[median_g > 1.7] <- 1.7
  
  
  #median_g[is.na(median_g)] <- min(median_g,na.rm=T)
  
  return(list('normalized_dubs'=median_g,
              'normalization_factor'=norm_factor_kept))
  
  
  #Quick correction
  #median_g[median_g < 0] <- 0
  #median_g[median_g > 2] <- 2
  
  
  #median_g[is.na(median_g)] <- min(median_g,na.rm=T)
  
  #return(median_g)
  
  #return(list('normalized_dubs'=median_g,
  #            'normalization_factor'=norm_factor_kept))
}



dub_df <- c()
#lag_df <- c()
for(drug in drugs){
  names_with_drug <- grep(drug,colnames(input_file),val=T)
  for(mating_type in mating_types){
    names_with_mating_type <- grep(mating_type,names_with_drug,val=T)
    
    tag_df <- c()
    
    for(tag in tags){
      names_with_tag <- grep(tag,names_with_mating_type,val=T)
      
      print(names_with_tag)
      
      if(length(names_with_tag) > 0){
        frequency_df <- c()
        count_df <- c()
        for (time_point in time_points) {
          #time_point_query <- paste(c('^',mating_type,time_point),collapse='')
          #names_with_time_point <- grep(time_point_query,names_with_tag,val=T)
          
          if (time_point == '0') {
            time_point_query <-
              paste(c(paste(c(
                'T0', mating_type
              ), collapse = '_'), '.*', tag), collapse = '')
            names_with_time_point <-
              grep(time_point_query, colnames(input_file), val = T)
          } else{
            time_point_query <- paste(c('^', mating_type, time_point), collapse = '')
            names_with_time_point <-
              grep(time_point_query, names_with_tag, val = T)
          }
          
          #print(names_with_time_poin)
          
          
          #names_with_time_point <- grep(time_point_query,names_with_mating_type,val=T)
          
          if (length(names_with_time_point) > 0) {
            counts <-
              apply(input_file[, names_with_time_point, drop = F], 1, sum)# + 1
            
            frequency_df <- cbind(frequency_df, counts / sum(counts))
            count_df <- cbind(count_df, counts)
            
            colnames(frequency_df)[ncol(frequency_df)] <-
              names_with_time_point[1]
          }
        }
        
        # stop()
        
        timepoints <- sapply(colnames(frequency_df), function(name) {
          name1 <- strsplit(name, split = mating_type)[[1]][2]
          
          name2 <- strsplit(name1, split = '_')[[1]][1]
          
          return(name2)
        })
        
        #par(mfrow=c(2,2))
        timepoint_names <- colnames(frequency_df)
        for (i in 1:(ncol(frequency_df) - 1)) {
          #main_name <- paste(c(timepoint_names[i+1],timepoint_names[i]),collapse = '')
          main_name <- c(timepoint_names[i + 1], timepoint_names[i])
          hist(log2(frequency_df[, i + 1] / frequency_df[, i]),
               main = main_name,
               breaks = 100)
        }
        
        # stop()
        
        
        #print(c(drug, mating_type, tag))
        dub_calc <- calculate_normalized_dubs(count_df + 1, timepoints)
        
       # stop()
        names(dub_calc) <- rownames(input_file)
        
        #corresponding_t0 <- grep(paste(c('T0',mating_type,tag),collapse='.*'),colnames(input_file),val=T)
        
        
        
        #t0_counts <- apply(input_file[,corresponding_t0,drop=F],1,sum)
        
        #dub_calc[t0_counts > 30 & is.na(dub_calc)] <- 0
        
        #dub_df_name <- paste(c(drug,mating_type,tag),collapse='_')
        
        #dub_df <- cbind(dub_df,dub_calc)
        # colnames(dub_df)[ncol(dub_df)] <- dub_df_name
        
        tag_df <- cbind(tag_df, dub_calc)
      #hist(dub_calc,breaks=100)
      }
    }
    
    dub_calc_both_tags <- apply(as.matrix(tag_df),1,mean,na.rm=T)
    
    corresponding_t0 <- grep(paste(c('T0',mating_type),collapse='_'),colnames(input_file),val=T)
    
    t0_counts <- input_file[,corresponding_t0,drop=F]
    
    #Set missing to 0
    fitness_to_0_criteria <- apply(t0_counts,1,max) > 100 & is.na(dub_calc_both_tags)
    dub_calc_both_tags[fitness_to_0_criteria] <- 0
    
    #Set negative to 0
    dub_calc_both_tags[dub_calc_both_tags < 0] <- 0
    
    #Max out fitness to 2x wildtpe
    dub_calc_both_tags[dub_calc_both_tags > 2] <- 2
    
    dub_df <- cbind(dub_df,dub_calc_both_tags)
    
    dub_df_name <- paste(c(drug,mating_type),collapse='_')
    colnames(dub_df)[ncol(dub_df)] <- dub_df_name
  }
}
dub_df <- as.data.frame(dub_df)



all_strains_A <- rownames(mapping_file)[grep('A',mapping_file[,1])]
corresponding_t0_A <- grep(paste(c('T0','A'),collapse='_'),colnames(input_file),val=T)
t0_counts_A <- input_file[,corresponding_t0_A,drop=F]
good_strains_A <- rownames(dub_df)[apply(t0_counts_A,1,max) > 100]
good_strains_A <- intersect(all_strains_A,good_strains_A)


all_strains_alpha <- rownames(mapping_file)[grep('alpha',mapping_file[,1])]
corresponding_t0_alpha <- grep(paste(c('T0','alpha'),collapse='_'),colnames(input_file),val=T)
t0_counts_alpha <- input_file[,corresponding_t0_alpha,drop=F]
good_strains_alpha <- rownames(dub_df)[apply(t0_counts_alpha,1,max) > 100]
good_strains_alpha <- intersect(all_strains_alpha,good_strains_alpha)



A_dub_df <- c()
alpha_dub_df <- c()

for(drug in drugs){
  for(mating_type in mating_types){
    name <- paste(c(drug,mating_type),collapse='_')
    if(mating_type == 'A'){
      A_dub_df <- cbind(A_dub_df,dub_df[good_strains_A,name])
      colnames(A_dub_df)[ncol(A_dub_df)] <- name
    }else{
      alpha_dub_df <- cbind(alpha_dub_df,dub_df[good_strains_alpha,name])
      colnames(alpha_dub_df)[ncol(alpha_dub_df)] <- name
    }
  }
}

A_dub_df <- as.data.frame(A_dub_df)
alpha_dub_df <- as.data.frame(alpha_dub_df)

#rownames(dub_df) <- 
#dub_df <- dub_df[,-4]

#dub_df