################################################################
################################################################
################################################################
########                                                ########
########                                                ########
########     Pipeline for analysis of XGA data and      ########
########     generation of all figures                  ########
########                                                ########
########                                                ########
################################################################
################################################################
################################################################

library(Cairo)
library(animation)
this.dir <- dirname(parent.frame(2)$ofile)
setwd(this.dir)

#Fixes bug with Mac OS update
dyn.load('/Applications/Jalview/jre/Contents/Home/lib/server/libjvm.dylib')


########
# Colour parameters
########
#
# Master colour scale for heatmap-style plots
my_color_list <- c(
  rgb(1, 0.45, 0.25),
  rgb(0.8, 0.25, 0.25),
  rgb(0, 0, 0),
  rgb(0.25, 0.45, 0.8),
  rgb(0.25, 0.75, 1)
)
# Gene colours for drawing legends
my_gene_colors <- list(
  'PDR5' = rgb(47, 144, 206, maxColorValue = 255),
  'SNQ2' = rgb(179, 231, 172, maxColorValue = 255),
  'YOR1' = rgb(233, 153, 76, maxColorValue = 255),
  'YBT1' = rgb(255, 255, 191, maxColorValue = 255),
  'YCF1' = rgb(166, 7, 13, maxColorValue = 255)#,
  #'BPT1' = rgb(32, 32, 32, maxColorValue = 255)
)
# Three and two gene colour scales generated by my_color_list
blue_black_orange <- grDevices::colorRampPalette(my_color_list)
black_blue <- grDevices::colorRampPalette(my_color_list[3:5])

blue_white_red <- grDevices::colorRampPalette(c('#FF0041',rgb(1,1,1),'#00A8FC'))

# Package containing main scripts
devtools::load_all('../packages/XGeneAnalysis')
devtools::document('../packages/XGeneAnalysis')


########
# Where to read/write files
########
#
# Raw data directory
input_data_directory <- '../data/'
# Processed data directory
output_data_directory <- '../data/output'
# Individual growth experiments are in their own subfolder
# in the input data directory
tecan_output_path <- 'tecan_analysis'
qrt_pcr_output_path <- 'qrt_pcr'
nn_output_dir <- 'nn_analysis'
# Genotyping files
all_twas_strains <- 'all_twas_strains.tsv'
extra_genotying_data <- 'extra_genotyping.tsv'
# Map of strain IDs to genotypes
mapping_filename <- 'twas_id_map_fixed.tsv'
# Previously-called genotyping accuracy
genotyping_accuracy_file <- 'leave-one-out-x-validation07192013.tsv'


########
# Run-specific configuration
########
#
# Stores parameters for separate runs
parameter_file <- 'sample_parameters.tsv'
# Sample name
sample <- 'X14.Nov'
#
# Parameters from this file are set below
#
# Read sample-specific variables
setwd(input_data_directory)
sample_pars <- read.table(parameter_file, stringsAsFactors = F)
sample_vals <- sample_pars[, sample]
# Set sample-specific variables from parameter file
# Sub-folder to create 'deliverables' (figures, data, etc)
output_results_directory <- sample_vals[1]
# What to call the run-specific resistances generated (deprecated)
resistance_output_prefix <- sample_vals[2]
# What to call the run-specific linear model
lm_output_file <- sample_vals[3]
# Processed sequencing data to read
sequencing_filename <- sample_vals[4]
# What to call the run-specific resistances generated (MATa)
A_resistance_filename <- sample_vals[5]
# What to call the run-specific resistances generated (MATalpha)
alpha_resistance_filename <- sample_vals[6]
# Name of file specifying what the solvent control is for each drug
# if 'none', default parameters are used for many functions
drug_control_pair <- sample_vals[7]
if (drug_control_pair == 'none') {
  drug_control_pair <- NULL
}

########
# Pre-run maintenance
########
#
# global control of what metric is used for analysis
#  'resistance' - compares growth in each drug to control
#  'growth' - takes growth in each drug as-is
growth_metric <- 'resistance'#growth'
# Create Sub directories for output if they don't exist
dir.create(output_data_directory, showWarnings = FALSE)
dir.create(output_results_directory, showWarnings = FALSE)
# Linear model input results, defaults to the same file as the output file
lm_input_file <- lm_output_file


########
# TASK CONTROL
########
#
# Specify which part of the analysis to run
#
# OPTIONS:
#
# 'all' - runs all tasks
#
# 'Create genotyping data' - processes initial
#   genotyping results to arrive at final set
#   of genotypes
#
# 'Create resistance metrics' - processes
#   sequencing data to arrive at resistance and growth
#   scores for each strain
##
# 'Genotyping Accuracy'
#   makes a plot evaluating genotyping accuracy and linkage
#
# 'Linear model'
#   creates genotype-phenotype associations using a
#   linear model, stores the results
#
# 'Pool Genotype Group Comparison'
#   groups strains by PDR5/SNQ2/YOR1/YBT1/YCF1/BPT1
#   genotype and compares group averages between
#   MATa and MATalpha pools
#
#
# 'Linear coefficient heatmap'
#   creates a heatmap of the generalized linear
#   modeling results
##
# 'Linear Fitness Landscape'
#   draws various 'linear' fitness landscapes
#
# 'Fitness Density Plot'
#   draws a density plot for resistances of different
#   groups (currently only fluconazole)
#
# 'Radial Fitness Landscape'
#   draws various 'radial' fitness landscapes
#
# 'Analyze Tecan Data'
#   all automated analysis of single-strain growth
#   assays

to_run <- c('Analyze Tecan Data')#Radial Fitness Landscape')#,'NN extensions')#'qRT-PCR' )#'Linear model')#'Fitness Density Plot')#'Merge mating types')#'Genotyping Accuracy')#)#)#'Create resistance metrics','Pool Genotype Group Comparison')


##Good as -is
if ('Create genotyping data' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  
  all_geno <- read.csv(
    all_twas_strains,
    head = T,
    sep = '\t',
    stringsAsFactors = F
  )
  picked_geno <- dplyr::filter(all_geno, Comment == 'Pick')
  
  #Wildtypes were re-genotyped
  picked_geno <- picked_geno[apply(picked_geno[, 11:26], 1, sum) != 0, ]
  
  #New genotyping data
  extra_strains <- read.csv(
    extra_genotying_data,
    head = T,
    sep = '\t',
    stringsAsFactors = F
  )
  picked_geno <- picked_geno[, colnames(extra_strains)]
  
  #Merge together
  final_set <- rbind(picked_geno, extra_strains)
  colnames(final_set)[2] <- 'Plate'
  write.table(
    final_set,
    file = mapping_filename,
    sep = '\t',
    quote = F,
    row.names = F
  )
}



##Good as-is
if ('Genotyping Accuracy' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  dir_name <- 'genotyping_accuracy'
  dir.create(paste(c(output_results_directory, dir_name), collapse = '/'), showWarnings = FALSE)
  #setwd(output_results_directory)
  #setwd(dir_name)
  filename <-
    paste(c(paste(
      c(output_results_directory, dir_name, 'per_gene_accuracy')
    )), collapse = '/')
  make_genotyping_accuracy_barplot(genotyping_accuracy_file, filename =
                                     filename)
  
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  A_genotyping_df <-
    A_genotyping_df[apply(A_genotyping_df[, 2:17], 1, sum) != 0, ]
  
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  alpha_genotyping_df <-
    alpha_genotyping_df[apply(alpha_genotyping_df[, 2:17], 1, sum) != 0, ]
  
  setwd(this.dir)
  filename <-
    paste(c(paste(
      c(output_results_directory, dir_name, 'linkage')
    )), collapse = '/')
  Cairo::CairoFonts(regular = "Arial:style=Regular",
                    bold = "Arial:style=Bold")
  CairoPDF(file = filename,
           width = 7.5,
           height = 6)
  make_linkage_map(A_genotyping_df, alpha_genotyping_df, color_map = blue_black_orange)
  dev.off()
  
  filename <-
    paste(c(paste(
      c(
        output_results_directory,
        dir_name,
        'knockout_distribution'
      )
    )), collapse = '/')
  CairoPDF(file = filename,
           width = 11.5,#7.5,
           height = 6)
  knockout_dist_barplot(mapping_file[apply(mapping_file[, 2:17], 1, sum) != 0, ])
  dev.off()
  
}


##Revised
if ('Create resistance metrics' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  input_file <- read.table(sequencing_filename,
                           head = T,
                           row.names = 1)
  mapping_file <- read.table(mapping_filename,
                             head = T,
                             row.names = 1)
  
  strains_in_both <- intersect(rownames(input_file),rownames(mapping_file))
  
  input_file <- input_file[strains_in_both, ]
  mapping_file <- mapping_file[strains_in_both, ]
  
  
  ##Do the calc
  drugs <- unique(sapply(colnames(input_file),function(name){
    strsplit(name,split='_')[[1]][2]
  }))
  drugs <- unique(drugs)
  drugs <- drugs[!(drugs %in% c('A','alpha'))]
  
  if(growth_metric == 'resistance'){
    if(is.null(drug_control_pair)){
      controls <- c('CTRL','CTRL2','CTRL3')
    }
    
    if (!is.null(drug_control_pair)) {
      controls <- list()
      drug_control_pair_file <- read.table(file = drug_control_pair)
      for (i in 1:nrow(drug_control_pair_file)) {
        #drug <- as.vector(drug_control_pair_file[i, 1])
        ctrl <- as.vector(drug_control_pair_file[i, 2])
        #drug_control_list[[drug]] <- ctrl
        controls <- c(controls, ctrl)
      }
    }
  }
  
  
  
  
  mating_types <- c('A','alpha')
  tags <- c('UP','DN')
  time_points <- c('5','10','15','20')
  
  ##Make a list for stuff
  relevant_list <- make_input_list(input_file,drugs)
  
  
  if(growth_metric == 'resistance'){
    drugs <- drugs[!(drugs %in% controls)]
  }
  
  all_fit_list <- list()
  log_fit_list <- list()
  for(drug in drugs){
    
    log_res <- calculate_log_resistance(relevant_list,drug, metric_returned = growth_metric)
    log_fit_list[[drug]] <- log_res
    
    all_fit_list[[drug]]$fits_A <- log_res$A_resistance
    all_fit_list[[drug]]$fits_alpha <- log_res$alpha_resistance
    
  }
  
  #Put into data frame
  A_df <- c()
  alpha_df <- c()
  for(drug in drugs[!grepl('CTRL',drugs)]){
    fit_A <- all_fit_list[[drug]]$fits_A
    A_df <- cbind(A_df,fit_A)
    colnames(A_df)[ncol(A_df)] <- paste(c(drug,'A'),collapse='_')
    
    fit_alpha <- all_fit_list[[drug]]$fits_alpha
    alpha_df <- cbind(alpha_df,fit_alpha)
    colnames(alpha_df)[ncol(alpha_df)] <- paste(c(drug,'alpha'),collapse='_')
  }
  
  genotyping_df <- mapping_file 
  
  A_df <- as.data.frame(A_df[!is.na(apply(A_df,1,function(x){sum(x)})),])
  alpha_df <- as.data.frame(alpha_df[!is.na(apply(alpha_df,1,function(x){sum(x)})),])
  A_resistance_file <- A_df
  alpha_resistance_file <- alpha_df
  
  setwd(this.dir)
  setwd(output_data_directory)

  A_filename <- paste(c(resistance_output_prefix,'_A.tsv'),collapse='')
  alpha_filename <- paste(c(resistance_output_prefix,'_alpha.tsv'),collapse='')
  
  
  write.table(A_df,sep='\t',quote=F,file=A_filename)
  write.table(alpha_df,sep='\t',quote=F,file=alpha_filename)
}

##New analysis - revised
if('Merge mating types' %in% to_run | 'all' %in% to_run){
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <-
    read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  
  A_resistance_file[A_resistance_file < 0] <- 0
  alpha_resistance_file[alpha_resistance_file < 0] <- 0
  
  #Shouldn't happen, but just in case
    
  A_genotyping_df <- mapping_file[rownames(A_resistance_file),]
  alpha_genotyping_df <- mapping_file[rownames(alpha_resistance_file),]
  
  
  single_genes <- c('PDR5','SNQ2','YBT1','YCF1','YOR1')
  
  combined_df <- cbind(A_genotyping_df[rownames(A_resistance_file),], A_resistance_file)
  split_df_A <- split_df_to_list(combined_df,single_genes)
  
  
  combined_df <- cbind(alpha_genotyping_df, alpha_resistance_file)
  split_df_alpha <- split_df_to_list(combined_df,single_genes)
  
  genes <- names(split_df_A)
  
  drugs <- sapply(colnames(A_resistance_file),function(name){strsplit(name,split='_')[[1]][1]})
  
  for(drug in drugs){
    comparison_df <- t(sapply(genes, function(name) {
      twas_frame_A <- split_df_A[[name]]
      
      twas_frame_alpha <- split_df_alpha[[name]]
      
      twas_metr_A <-
        mean(twas_frame_A[, sprintf('%s_%s', drug, 'A')], na.rm = T)
      twas_metr_alpha <-
        mean(twas_frame_alpha[, sprintf('%s_%s', drug, 'alpha')], na.rm = T)
      
      
      
      return(c(twas_metr_A, twas_metr_alpha))
    }))
    
    colnames(comparison_df) <- c('MATa', 'MATalpha')
    comparison_df <- as.data.frame(comparison_df)
    
    comparison_lm <- lm(MATalpha ~ MATa, data = comparison_df)
    
    print(comparison_lm)
    
    coefs <- coefficients(comparison_lm)
    
    drug_A <- paste(c(drug, 'A'), collapse = '_')
    
    A_resistance_file[, drug_A] <-
      A_resistance_file[, drug_A] * coefs[2] + coefs[1]
    
  }
 
  colnames(A_resistance_file) <- sapply(drugs,function(drug){paste(c(drug,'both'),collapse = '_')})
  colnames(alpha_resistance_file) <- sapply(drugs,function(drug){paste(c(drug,'both'),collapse = '_')})
  
  combined_resistance_file <- rbind(A_resistance_file,alpha_resistance_file)
  combined_resistance_file <- as.data.frame(combined_resistance_file)
  
  filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  #alpha_filename <- paste(c(resistance_output_prefix,'_alpha.tsv'),collapse='')
  
  #write.table(A_df,sep='\t',quote=F,file=A_filename)
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  #stop()
  
  write.table(combined_resistance_file ,sep='\t',quote=F,file=filename)
  
}


##Revised
if ('Linear model' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  
  #Shouldn't happen, but just in case
  A_resistance_file <-
    A_resistance_file[!is.na(apply(A_resistance_file, 1, sum)), ]
  alpha_resistance_file <-
    alpha_resistance_file[!is.na(apply(alpha_resistance_file, 1, sum)), ]
  
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-  mapping_file[rownames(alpha_resistance_file), ]
  
  
  A_resistance_file[A_resistance_file < 1e-10] <- 1e-10
  alpha_resistance_file[alpha_resistance_file < 1e-10] <- 1e-10
  
  
  
  lm_results_single <- create_all_lms(
    A_resistance_file,
    alpha_resistance_file,
    A_genotyping_df,
    alpha_genotyping_df,
    lm_method = 'glm',
    reduce_features_by_glm = F,
    maximum_degree = 1
  )
  
  
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  both_res_file[both_res_file < 1e-10] <- 1e-10
  mapfile <- mapping_file[rownames(both_res_file),]
  

  
  lm_results_full <- build_complex_glms(mapfile, both_res_file, maximum_complexity = 5)
  
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  save(list = c('lm_results_full', 'lm_results_single'),
       file = lm_output_file)
  lm_results_list <- list(marginal_search = lm_results_single,
                          full_search = lm_results_full)
  write_results_to_xlsx(lm_results_list, 'lm_result_summary.xlsx')
}


if ('Pool Genotype Group Comparison' %in% to_run |
    'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename,
                             head = T,
                             row.names = 1)
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  
  
  A_resistance_file[A_resistance_file < 0] <- 0
  alpha_resistance_file[alpha_resistance_file < 0] <- 0
  
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  
  
  setwd(this.dir)
  outdir <-
    paste(c(output_results_directory, 'group_comparison_scatterplots'),
          collapse = '/')
  dir.create(outdir, showWarnings = FALSE)
  
  drugs <-
    sapply(colnames(alpha_resistance_file), function(name) {
      strsplit(name, split = '_')[[1]][1]
    })
  
  Cairo::CairoFonts(regular = "Arial:style=Regular",
                    bold = "Arial:style=Bold")
  CairoPDF(paste(
    c(outdir, 'group_reproducibility_histogram.pdf'),
    collapse = '/'
  ),
  width = 4.2,
  height = 4.4)
  
  mata_matalpha_comparison_histogram(
    A_resistance_file = A_resistance_file,
    alpha_resistance_file = alpha_resistance_file,
    A_genotyping_df = A_genotyping_df,
    alpha_genotyping_df = alpha_genotyping_df,
    genes = names(my_gene_colors)[1:5],
    xlims = c(0,1),
    breaks = 10
  )
  dev.off()
  
  for (drug in drugs) {
    Cairo::CairoFonts(regular = "Arial:style=Regular",
                      bold = "Arial:style=Bold")
    CairoPDF(paste(c(
      outdir, sprintf('group_reproducibility_%s.pdf', drug)
    ), collapse = '/'),
    width = 6,
    height = 6)
    mata_matalpha_comparison_scatterplot(
      A_resistance_file = A_resistance_file,
      alpha_resistance_file = alpha_resistance_file,
      A_genotyping_df = A_genotyping_df,
      alpha_genotyping_df = alpha_genotyping_df,
      drug = drug,
      gene_palette = my_gene_colors[1:5],
      nsample_cutoff = 1,
      xlims = c(-0.3, 1.5),
      ylims = c(-0.3, 1.5)
    )
    dev.off()
    
    #stop()
  }
}


##Under revision
if ('Linear coefficient heatmap' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(output_data_directory)
  #This loads the 'lm_results' variable
  load(lm_output_file)
  
  
  #Analyze max genetic interaction complexity for each gene
  complexities <- list()
  drugs <- names(lm_results_full$term_names)
  for(drug in drugs){
    coefs <- lm_results_full[['term_names']][[drug]][['both']]
    
    for(coef in coefs){
      genes <- strsplit(coef,split=':')[[1]]
      for(gene in genes){

        complexities[[gene]] <- max(c(complexities[[gene]],length(genes)))
      }
    }
  }
  
  #First we analyze the single results to report some numbers in the text (no output)
  overlaps <- lm_results_single$overlap_results
  
  existing_associations <- read.xlsx2('../../doc/additional files/Data_S7_drug-knockout-associations.xlsx', sheetIndex = 1)
  existing_result_list <- list()
  for(i in 1:nrow(existing_associations)){
    row <- existing_associations[i,]
    row <- as.vector(unlist(row))
    if(is.null(existing_result_list[[row[1]]])){
      existing_result_list[[row[1]]] <- list()
    }
    existing_result_list[[row[1]]][[row[2]]] <- row[3]
  }
  
  
  all_result_list <- list()
  all_weak_result_list <- list()
  all_strong_result_list <- list()
  
  all_previous_result_list <- list()
  
  novel_all_result_list <- list()
  novel_weak_result_list <- list()
  novel_strong_result_list <- list()
  
  strong_threshold <- 0.9
  
  
  for(drug in names(overlaps)){
    for(gene in overlaps[[drug]]){
      effect_size_A <- lm_results_single$lm_list[[drug]][['A']]$coefficients[[gene]]
      effect_size_alpha <- lm_results_single$lm_list[[drug]][['alpha']]$coefficients[[gene]]
      effect_size <- mean(c(effect_size_A,effect_size_alpha))
      
      all_result_list[[drug]][[gene]] <- effect_size
      
      if(abs(effect_size) > -log(strong_threshold)){
        all_strong_result_list[[drug]][[gene]] <- effect_size
      }else{
        all_weak_result_list[[drug]][[gene]] <- effect_size
      }
      
      if(is.null(existing_result_list[[drug]][[gene]])){
        novel_all_result_list[[drug]][[gene]] <- effect_size
        if(abs(effect_size) > -log(strong_threshold)){
          novel_strong_result_list[[drug]][[gene]] <- effect_size
        }else{
          novel_weak_result_list[[drug]][[gene]] <- effect_size
        }
        
      }else {
        if((effect_size > 0 & existing_result_list[[drug]][[gene]] == 'sensitivity') |
           (effect_size < 0 & existing_result_list[[drug]][[gene]] == 'resistance')){
          novel_all_result_list[[drug]][[gene]] <- effect_size
          if(abs(effect_size) > -log(strong_threshold)){
            novel_strong_result_list[[drug]][[gene]] <- effect_size
          }else{
            novel_weak_result_list[[drug]][[gene]] <- effect_size
          }
        }else{
          all_previous_result_list[[drug]][[gene]] <- effect_size
        }
      }
      
    }
  }
  
  setwd(this.dir)
  dir.create(paste(c(output_results_directory, 'heatmaps'), collapse = '/'), showWarnings = FALSE)
  
  Cairo::CairoFonts(regular = "Arial:style=Regular", italic="Arial:style=Italic")
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap.pdf'
      ),
      collapse = '/'
    ),
    width = 17,
    height = 7
  )
  Cairo::CairoFonts(regular = "Arial:style=Regular", italic="Arial:style=Italic")
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    draw_complexity_legend = F,
    text_size = 1,
    na.color = rgb(0.4, 0.4, 0.4)
  )
  dev.off()
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_for_three_drugs.pdf'
      ),
      collapse = '/'
    ),
    width = 12,
    height = 5.5
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('cisplatin','bisantrene','mitoxantrone'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.06,
    reorder_drugs = F,
    draw_complexity_legend = T,
    split_complexity_by_line = T
  )
  dev.off()
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_cisplatin.pdf'
      ),
      collapse = '/'
    ),
    width = 15,
    height = 9
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('cisplatin'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.06,
    reorder_drugs = F,
    draw_complexity_legend = F,
    split_complexity_by_line = F,
    text_size = 2.7,
    legend_text_size = 2.7,
    margins = c(37,7,5,11)#,
    #text_angle = 90
  )
  dev.off()
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_mitoxantrone.pdf'
      ),
      collapse = '/'
    ),
    width = 14,
    height = 9
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('mitoxantrone'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.08,
    reorder_drugs = F,
    draw_complexity_legend = F,
    split_complexity_by_line = F,
    text_size = 2.7,
    legend_text_size = 2.7,
    margins = c(37,13,5,12)#,
    #text_angle = 90
  )
  dev.off()
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_bisantrene.pdf'
      ),
      collapse = '/'
    ),
    width = 17,
    height = 9
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('bisantrene'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.08,
    reorder_drugs = F,
    draw_complexity_legend = F,
    split_complexity_by_line = F,
    text_size = 2.7,
    legend_text_size = 2.7,
    margins = c(37,13,5,12)#,
    #text_angle = 90
  )
  dev.off()
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_fluconazole.pdf'
      ),
      collapse = '/'
    ),
    width = 14,
    height = 9
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('fluconazole'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.08,
    reorder_drugs = F,
    draw_complexity_legend = F,
    split_complexity_by_line = F,
    text_size = 2.7,
    legend_text_size = 2.7,
    margins = c(37,13,5,12)#,
    #text_angle = 90
  )
  dev.off()
  
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_benomyl.pdf'
      ),
      collapse = '/'
    ),
    width = 12,
    height = 7.5
  )
  lm_coefficient_heatmap_v2(
    lm_results_full,
    blue_white_red,
    drugs = c('benomyl'),
    dispersion_constant = 1,
    #max_complexity_plotted = 4,
    scale_offset = 0.08,
    reorder_drugs = F,
    draw_complexity_legend = F,
    split_complexity_by_line = F,
    text_size = 2.7,
    legend_text_size = 2.7,
    margins = c(30,10,5,12)#,
    #text_angle = 90
  )
  dev.off()
  
  
  Cairo::CairoPDF(
    file = paste(
      c(
        output_results_directory,
        'heatmaps',
        'linear_coefficient_heatmap_single.pdf'
      ),
      collapse = '/'
    ),
    width = 9,
    height = 6
  )
  Cairo::CairoFonts(regular = "Arial:style=Regular")
  lm_coefficient_heatmap(
    lm_results_single,
    blue_white_red,
    lwid = c(1, 0.01, 0.2, 5),
    breaks = c(-25:25) / 100,
    lhei = c(0.5, 4),
    na.color = rgb(0.4,0.4,0.4)
  )
  dev.off()
}




##Revised
if ('Linear Fitness Landscape' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename,
                             head = T,
                             row.names = 1)
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  
  
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  mapfile <- mapping_file[rownames(both_res_file),]
  
  
  drugs <- sapply(colnames(both_res_file),function(x){strsplit(x,split='_')[[1]][1]})
  
  setwd(this.dir)
  setwd(output_results_directory)
  dir.create('fitness_landscape_graphs', showWarnings = F)
  
    
    for(drug in drugs){
      setwd(this.dir)
      setwd(output_results_directory)
      
      Cairo::CairoFonts(regular = "Arial:style=Regular",
                        bold = "Arial:style=Bold")
      
      file_name <- sprintf('%s_custom_aspect', drug)
      
      Cairo::CairoPDF(file = paste(c('fitness_landscape_graphs/', file_name, '.pdf'), collapse = ''),
                      width = 11,
                      height = 9.85)
  
      exhaustive_landscape_drawing(
        both_res_file,
        alpha_resistance_file = NULL,
        mapfile,
        alpha_genotyping_df = NULL,
        gene_colors = my_gene_colors,
        drawn_genes = names(my_gene_colors),
        drugs = drug,
        legend = F,
        mating_types = 'both',
        sidebyside = F,
        separate_files = T,
        make_pdf = F,
        box_height = 0.04416,
        box_width = 0.0333
      )
      dev.off()
      
      #Make normalized fitness landscape for fluconazole (Figure 6)
      if(drug == 'fluconazole'){
        both_res_file$fluconazole_both <-
          both_res_file$fluconazole_both / max(both_res_file$fluconazole_both)
        
        file_name <- sprintf('%s_custom_aspect_normalized', drug)
        
        Cairo::CairoPDF(file = paste(c('fitness_landscape_graphs/', file_name, '.pdf'), collapse = ''),
                        width = 11,
                        height = 9.85)
        exhaustive_landscape_drawing(
          both_res_file,
          alpha_resistance_file = NULL,
          mapfile,
          alpha_genotyping_df = NULL,
          gene_colors = my_gene_colors,
          drawn_genes = names(my_gene_colors),
          drugs = drug,
          legend = F,
          mating_types = 'both',
          sidebyside = F,
          separate_files = T,
          make_pdf = F,
          box_height = 0.04416,
          box_width = 0.0333
        )
        dev.off()
        
        
        
      }
      
    }
}

if ('Fitness Density Plot' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  setwd(this.dir)
  setwd(output_data_directory)
  
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  
  combined_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  
  combined_resistance_file <- read.table(combined_filename)
  combined_genotyping_df <- mapping_file[rownames(combined_resistance_file), ]
  
  
  setwd(this.dir)
  setwd(output_results_directory)
  dir.create('split_density_distribution', showWarnings = FALSE)
  setwd('split_density_distribution')
  
  
  seven_gene_palette = c(my_gene_colors,list('BPT1' = rgb(32, 32, 32, maxColorValue = 255),
                                             'VMR1' = rgb(187, 110, 255, maxColorValue = 255)))
  
  
  pdf('bisantrene_density.pdf', width = 3, height = 6)
  draw_split_density_distribution(
    A_genotyping_df = combined_genotyping_df,
    A_resistance_file = combined_resistance_file,
    drug = 'bisantrene',
    mating_type = 'both',
    gene_palette = seven_gene_palette,
    ylims=c(-0.1,1.1),
    split_groups_by = 'PDR5',
    density_colors = c(rgb(77/255, 77/255, 77/255, 0.3), rgb(17/255, 48/255, 127/255, 0.3)),
    genes = c("SNQ2", "YOR1","YBT1","VMR1","YCF1"),
    genes_to_split_by  = c("SNQ2", "YOR1","YBT1","YCF1","VMR1"),
    point_size = 1,
    groups_drawn = c("wt","VMR1","SNQ2:YBT1:YCF1:YOR1","SNQ2:VMR1:YBT1:YCF1:YOR1"),
    swarm_size = 0.01
  )
  dev.off()
  
  pdf('mitoxantrone_density.pdf', width = 3, height = 4.5)
  draw_split_density_distribution(
    A_genotyping_df = combined_genotyping_df,
    A_resistance_file = combined_resistance_file,
    drug = 'mitoxantrone',
    mating_type = 'both',
    gene_palette = seven_gene_palette,
    ylims=c(0.8,1.03),
    split_groups_by = 'none',
    density_colors = c(rgb(77/255, 77/255, 77/255, 0.3), rgb(17/255, 48/255, 127/255, 0.3)),
    genes = c("PDR5","YOR1","YCF1","SNQ2","BPT1"),
    genes_to_split_by = c("PDR5","YOR1","YCF1","SNQ2","BPT1"),
    point_size = 1,
    groups_drawn = c("wt","BPT1","PDR5:SNQ2:YCF1:YOR1","BPT1:PDR5:SNQ2:YCF1:YOR1"),
    swarm_size = 0.01
  )
  dev.off()
  
  pdf('cisplatin_density_bpt1.pdf', width = 3, height = 4.5)
  draw_split_density_distribution(
    A_genotyping_df = combined_genotyping_df,
    A_resistance_file = combined_resistance_file,
    drug = 'cisplatin',
    mating_type = 'both',
    gene_palette = seven_gene_palette,
    ylims=c(0.6,1.25),
    split_groups_by = 'none',
    density_colors = c(rgb(77/255, 77/255, 77/255, 0.3), rgb(17/255, 48/255, 127/255, 0.3)),
    genes = c("PDR5","YOR1","YCF1","SNQ2","BPT1"),
    genes_to_split_by = c("PDR5","YOR1","YCF1","SNQ2","BPT1"),
    point_size = 1,
    groups_drawn = c("wt","BPT1","PDR5:SNQ2:YCF1:YOR1","BPT1:PDR5:SNQ2:YCF1:YOR1"),
    swarm_size = 0.01
  )
  dev.off()
  
  pdf('bisantrene_density.pdf', width = 4, height = 4.5)
  draw_split_density_distribution(
    A_genotyping_df = combined_genotyping_df,
    A_resistance_file = combined_resistance_file,
    drug = 'bisantrene',
    mating_type = 'both',
    gene_palette = seven_gene_palette,
    ylims=c(-0.1,1.1),
    split_groups_by = 'none',
    density_colors = c(rgb(77/255, 77/255, 77/255, 0.3), rgb(17/255, 48/255, 127/255, 0.3)),
    genes = c("SNQ2", "YOR1","YBT1","VMR1","YCF1"),
    genes_to_split_by  = c("SNQ2", "YOR1","YBT1","YCF1","VMR1"),
    point_size = 1,
    groups_drawn = c("wt","VMR1","YBT1","VMR1:YBT1","SNQ2:YBT1:YCF1:YOR1","SNQ2:VMR1:YBT1:YCF1:YOR1"),
    swarm_size = 0.01,
    genotype_circle_width_constant = 0.06
  )
  dev.off()
  
  
}

if ('Radial Fitness Landscape' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  setwd(this.dir)
  setwd(output_data_directory)
  
  
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  
  setwd(this.dir)
  setwd(output_data_directory)
  
  both_res_file <- read.table(both_res_filename)
  mapping_file <- mapping_file[rownames(both_res_file),]
  
  A_resistance_file <- both_res_file[grep('alpha',mapping_file$Plate,invert = T),]
  alpha_resistance_file <- both_res_file[grep('alpha',mapping_file$Plate,invert = F),]
  
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  
  
  colnames(A_resistance_file) <- sapply(colnames(A_resistance_file),function(name){paste(c(strsplit(name,split='_')[[1]][1],'_A'),collapse='')})
  colnames(alpha_resistance_file) <- sapply(colnames(alpha_resistance_file),function(name){paste(c(strsplit(name,split='_')[[1]][1],'_alpha'),collapse='')})
  
  ctrl_list <- list(
    'all' = c(
      #'benomyl',
      'camptothecin',
      #'tamoxifen',
      'methotrexate',
      'ketoconazole',
      'cycloheximide',
      #'ketoconazole',
      'fluconazole',
      #'beauvericin',
      'bisantrene',
      'valinomycin',
      'mitoxantrone'
      
    ),
    'all_presentation_ver' = c(
      'benomyl',
      'camptothecin',
      'tamoxifen',
      'methotrexate',
      'cisplatin',
      'ketoconazole',
      'fluconazole',
      'beauvericin',
      'bisantrene',
      'valinomycin',
      'mitoxantrone',
      'cycloheximide'
    ),
    'other' = c(
      #'fluconazole',
      'cisplatin',
      'itraconazole',
      'miconazole',
      'colchicine',
      'imatinib',
      'tamoxifen',
      'beauvericin'
    )
  )
  
  for (i in 1:length(ctrl_list)) {
    setwd(this.dir)
    setwd(output_results_directory)
    
    exhaustive_radial_landscape_drawing(
      A_resistance_file,
      alpha_resistance_file,
      A_genotyping_df,
      alpha_genotyping_df,
      color_function = blue_black_orange,
      drugs = ctrl_list[[i]],
      drawn_genes = names(my_gene_colors),
      filename = names(ctrl_list[i]),
      dpi = 50,
      #drawn_genes=names(my_gene_colors),
      color_map = blue_black_orange,
      nrows_all_genes = ifelse(names(ctrl_list)[i] %in% c('all','other'), 4, 3),
      mating_types = c('A', 'alpha'),
      sig_threshold_marginal = 0.05,
      maximum_degree = 4,
      border_color = rgb(0.1, 0.1, 0.1, 0.1),
      sidebyside = F,
      all_drugs_combined = T,
      draw_title = F,
      keep_title_margins = T,
      output_results_directory = 'fitness_landscape_graphs_radial',
      p_cutoff = 1e09
    )
  }
  plot_control_list <-
    list(
      list(
        drug = 'benomyl',
        depth = 1,
        mating_type = 'A',
        plot_width = 9.6 / 3,
        plot_height = 9.6 / 3
      ),
      list(
        drug = 'benomyl',
        depth = 2,
        mating_type = 'A',
        plot_width = 9.6 / 2,
        plot_height = 9.6 / 2
      ),
      list(
        drug = 'benomyl',
        depth = 5,
        mating_type = 'A',
        plot_width = 9.6,
        plot_height = 9.6
      ),
      list(
        drug = 'benomyl',
        depth = 5,
        mating_type = 'alpha',
        plot_width = 9.6,
        plot_height = 9.6
      ),
      
      list(
        drug = 'fluconazole',
        depth = 5,
        mating_type = 'A',
        plot_width = 9.6,
        plot_height = 9.6
      ),
      list(
        drug = 'fluconazole',
        depth = 5,
        mating_type = 'alpha',
        plot_width = 9.6,
        plot_height = 9.6
      ),
      list(
        drug = 'valinomycin',
        depth = 5,
        mating_type = 'alpha',
        plot_width = 9.6,
        plot_height = 9.6
      )
    )
  
  for (i in 1:length(plot_control_list)) {
    control <- plot_control_list[[i]]
    drug <- control$drug
    depth <- control$depth
    plot_width <- control$plot_width
    plot_height <- control$plot_height
    mating_type <- control$mating_type
    
    setwd(this.dir)
    setwd(output_results_directory)
    
    exhaustive_radial_landscape_drawing(
      A_resistance_file,
      alpha_resistance_file,
      A_genotyping_df,
      alpha_genotyping_df,
      color_function = blue_black_orange,
      drugs = drug,
      drawn_genes = names(my_gene_colors),
      dpi = 150,
      #drawn_genes=c('YBT1','PDR10','PDR12','PDR5','SNQ2','YOR1'),
      color_map = blue_black_orange,
      #nrows_all_genes = 3,
      mating_types = mating_type,
      #,'alpha'),
      sig_threshold_marginal = 0.05,
      maximum_degree = 4,
      border_color = rgb(0.1, 0.1, 0.1, 0.1),
      sidebyside = T,
      all_drugs_combined = T,
      output_results_directory = 'fitness_landscape_graphs_radial',
      filename = sprintf('%s_%s_knockouts_%s', drug, depth, mating_type),
      max_depth = depth,
      nrows_all_genes = 1,
      plot_width = plot_width,
      plot_height = plot_height,
      draw_mode = 'pdf',
      circle_resolution = 100,
      draw_title = F,
      p_cutoff = 1e09
    )
  }
  
}

if ('Analyze Tecan Data' %in% to_run | 'all' %in% to_run) {
  setwd(this.dir)
  tecan_output_path <-
    paste(c(output_results_directory, tecan_output_path), collapse = '/')
  dir.create(tecan_output_path, showWarnings = FALSE)
  summary_output_path = paste(c(tecan_output_path, 'experiment_summary_plots'),
                              collapse = '/')
  dir.create(summary_output_path, showWarnings = FALSE)
  
  genotypes_tested <- c('PDR5', 'SNQ2', 'YOR1', 'YBT1', 'YCF1')
  correspondence_dir <- 'all'
  setwd(this.dir)
  setwd(paste(
    c(input_data_directory, 'marinella/', correspondence_dir),
    collapse = ''
  ))
  file_correspondence_list <- 'file_correspondence_list.tsv'
  
  
  od_list <- create_od_list(file_correspondence_list)
  
  drugs <- names(od_list)[names(od_list) != 'dmso']

  
  drug <- 'fluconazole'
  experiment_summary_df <- tecan_od_list_summary(od_list, drug, names(od_list[[drug]]))
  setwd(this.dir)
  setwd(output_data_directory)
  write.table(experiment_summary_df,sep='\t',row.names = F, quote = F, 'fluconazole_resistance_tecan.tsv')
  
  # for (drug in 'fluconazole') {
  #   for (conc in names(od_list[[drug]])) {
  #     plot_experiment_summary(
  #       od_list,
  #       drug = drug,
  #       drug_conc = conc,
  #       control = 'dmso',
  #       control_conc = '2%',
  #       draw_pdf = T,
  #       parent_dir = this.dir,
  #       output_dir = summary_output_path,
  #       max_od = 1.5,
  #       genotypes_tested = genotypes_tested
  #     )
  #   }
  # }
  # 
  
  #We will compare XGA pool data with the tecan data
  setwd(this.dir)
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  A_resistance_file <- read.table(A_resistance_filename)
  alpha_resistance_file <- read.table(alpha_resistance_filename)
  
  A_genotyping_df <- mapping_file[rownames(A_resistance_file), ]
  alpha_genotyping_df <-
    mapping_file[rownames(alpha_resistance_file), ]
  
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  both_res_file[both_res_file < 1e-10] <- 1e-10
  both_res_file <- apply(both_res_file,2,function(x){x/max(x)})
  
  mapfile <- mapping_file[rownames(both_res_file),]
  
  #Quick fix - original averaged between A and alpha
  #but better to put them in combined
  #and average that
  #just going to make average the combined data twice (same as averaging one)
  #instead of changing function itself for now
  A_resistance_file_hack <- both_res_file
  alpha_resistance_file_hack <- both_res_file
  A_genotyping_df_hack <- mapfile
  alpha_genotyping_df_hack <- mapfile
  
  colnames(A_resistance_file_hack) <- colnames(A_resistance_file)
  colnames(alpha_resistance_file_hack) <- colnames(alpha_resistance_file)
  
  
  setwd(this.dir)
  for (conc in names(od_list$fluconazole)) {
    CairoPDF(paste(c(
      tecan_output_path,
      sprintf('twas_reproducibility_%s.pdf', paste(c(
        'fluconazole', conc
      ), collapse = '_'))
    ), collapse = '/'),
    width = 6,
    height = 6)
    twas_tecan_comparison_scatterplot(
      od_list,
      A_resistance_file_hack,
      alpha_resistance_file_hack,
      A_genotyping_df_hack,
      alpha_genotyping_df_hack,
      drug = 'fluconazole',
      concentration = conc,
      control = 'dmso',
      control_conc = '2%',
      gene_palette = list(
        'PDR5' = 
          rgb(47, 144, 206, maxColorValue = 255),
        'SNQ2' =
          rgb(179, 231, 172, maxColorValue = 255),
        'YOR1' =
          rgb(233, 153, 76, maxColorValue = 255),
        'YCF1' =
          rgb(166, 7, 13, maxColorValue = 255),
        'YBT1' =
          rgb(255, 255, 191, maxColorValue = 255)
      )
    )
    dev.off()
  }
  
  setwd(this.dir)
  CairoPDF(paste(c(
    tecan_output_path,
    'twas_vs_ic50_fluconazole.pdf'
  ), collapse = '/'),
  width = 6,
  height = 6)
  twas_tecan_comparison_scatterplot(
    od_list,
    A_resistance_file_hack,
    alpha_resistance_file_hack,
    A_genotyping_df_hack,
    alpha_genotyping_df_hack,
    drug = 'fluconazole',
    metric = 'IC50',
    control = 'dmso',
    control_conc = '2%',
    gene_palette = list(
      'PDR5' = 
        rgb(47, 144, 206, maxColorValue = 255),
      'SNQ2' =
        rgb(179, 231, 172, maxColorValue = 255),
      'YOR1' =
        rgb(233, 153, 76, maxColorValue = 255),
      'YCF1' =
        rgb(166, 7, 13, maxColorValue = 255),
      'YBT1' =
        rgb(255, 255, 191, maxColorValue = 255)
    ))
  dev.off()
}

if('NN optimization' %in% to_run | 'all' %in% to_run){
  set.seed(123)
  setwd(this.dir)
  
  
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  
  #Scale to 0-1 interval
  both_res_file[both_res_file < 1e-10] <- 1e-10
  both_res_file <- apply(both_res_file,2,function(x){x/max(x)})
  
  mapfile <- mapping_file[rownames(both_res_file),]
  
  genes <- colnames(mapfile)[2:17]
  condition_names <- colnames(both_res_file)
  drugs <- sapply(colnames(both_res_file),function(x){strsplit(x,split='_')[[1]][1]})
  
  alpha_ind <- grep('alpha',mapfile$Plate)
  A_ind <- grep('alpha',mapfile$Plate, invert = T)
  
  ranges_tested = c('-3_-4_11','-6_0_13')
  
  for(range_tested in ranges_tested) {
    range_tested <- strsplit(range_tested,split='_')[[1]]
    
    regs <-
      10 ^ seq(as.numeric(range_tested[1]),
               as.numeric(range_tested[2]),
               length.out = as.numeric(range_tested[3]))
    performance_alpha <-
      nn_model_search(
        both_res_file[alpha_ind, ],
        mapfile[alpha_ind, ],
        condition_names,
        genes,
        regularization_rates = regs[1:3],
        epochs = 1000,
        batch_size = round(0.3 * length(alpha_ind))
      )
    
    performance_A <-
      nn_model_search(
        both_res_file[A_ind, ],
        mapfile[A_ind, ],
        condition_names,
        genes,
        regularization_rates = regs,
        epochs = 10000,
        batch_size = round(0.3 * length(A_ind))
      )
    performance_both <-
      nn_model_search(
        both_res_file,
        mapfile,
        condition_names,
        genes,
        regularization_rates = regs,
        epochs = 10000,
        batch_size = round(0.3 * nrow(both_res_file))
      )
    
    
    nn_output_path <-
      paste(c(output_results_directory, nn_output_dir), collapse = '/')
    setwd(this.dir)
    setwd(input_data_directory)
    dir.create(nn_output_path, showWarnings = FALSE)
    setwd(nn_output_path)
    
    filename <- sprintf('performance_data_10_%s_%s',range_tested[1],range_tested[2])
    
    save(
      list = c('performance_alpha',
               'performance_A',
               'performance_both'),
      file = filename
    )
    
    
    both_perf <- c(performance_A[, 2], performance_alpha[, 2])
    both_param <- c(performance_A[, 4], performance_alpha[, 4])
    
    
    filename <- sprintf('mse_vs_reg_rate_%s_%s.pdf',range_tested[1],range_tested[2])
    
    pdf(file = filename,
        width = 4.5,
        height = 3.5)
    par(mar = c(4.5, 4.5, 1, 1))
    plot(
      log10(performance_alpha[, 1]),
      (performance_alpha[, 2]),
      type = 'l',
      col = 'red',
      ylim = c(min(both_perf),
               max(both_perf)),
      xlab = expression(paste("log" ["10"], " (Regulariation rate)")),
      ylab = 'Mean Squared Error',
      cex.lab = 1.5
    )
    lines(log10(performance_A[, 1]), (performance_A[, 2]), col = 'blue')
    abline(v = -4, lty = 2)
    legend(
      -2.5,
      quantile(both_perf, probs = 0.29),
      legend = c('MATa',
                 expression(paste('MAT', alpha))),
      fill = c('blue', 'red')
    )
    dev.off()
    
    filename <- sprintf('mse_vs_reg_nparam_%s_%s.pdf',range_tested[1],range_tested[2])
    
    pdf(file = filename,
        width = 4.5,
        height = 3.5)
    par(mar = c(4.5, 4.5, 1, 1))
    plot(
      log10(performance_alpha[, 1]),
      (performance_alpha[, 4]),
      type = 'l',
      col = 'red',
      ylim = c(min(both_param),
               max(both_param)),
      xlab = expression(paste("log" ["10"], " (Regulariation rate)")),
      ylab = 'Number of Parameters',
      cex.lab = 1.5
    )
    lines(log10(performance_A[, 1]), (performance_A[, 4]), col = 'blue')
    legend(
      -2.5,
      quantile(both_param, probs = 0.8),
      legend = c('MATa',
                 expression(paste('MAT', alpha))),
      fill = c('blue', 'red')
    )
    abline(v = -4, lty = 2)
    dev.off()
  }
}

if('NN analysis' %in% to_run | 'all' %in% to_run){
  set.seed(123)
  setwd(this.dir)
  
  
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  
  #Scale to 0-1 interval, with small lower limit
  both_res_file[both_res_file < 1e-10] <- 1e-10
  both_res_file <- apply(both_res_file,2,function(x){x/max(x)})
  
  mapfile <- mapping_file[rownames(both_res_file),]
  
  genes <- colnames(mapfile)[2:17]
  condition_names <- colnames(both_res_file)
  drugs <- sapply(colnames(both_res_file),function(x){strsplit(x,split='_')[[1]][1]})
  
  alpha_ind <- grep('alpha',mapfile$Plate)
  A_ind <- grep('alpha',mapfile$Plate, invert = T)
  
  A_filter <- !(apply(mapfile[A_ind,2:17],1,paste,collapse='') %in% apply(mapfile[alpha_ind,2:17],1,paste,collapse=''))
  alpha_filter <- !(apply(mapfile[alpha_ind,2:17],1,paste,collapse='') %in% apply(mapfile[A_ind,2:17],1,paste,collapse=''))
  
  A_ind <- A_ind[A_filter]
  alpha_ind <- alpha_ind[alpha_filter]
  
  
  nn_output_path <-
    paste(c(output_results_directory,nn_output_dir), collapse = '/')
  setwd(this.dir)
  setwd(input_data_directory)
  dir.create(nn_output_path, showWarnings = FALSE)
  setwd(nn_output_path)
  
  #Hyperparameters for neural network
  learning_rate <- 0.05
  epochs <- 10000
  regularization_rate <- 5e-04
  
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  if('sig_features_alpha_16_drugs.RData' %in% list.files()){
    load('sig_features_alpha_16_drugs.RData')
    nn_alpha <- merge_many_nn_models(resfile = both_res_file[alpha_ind, ],
                                     mapfile = mapfile[alpha_ind, ],
                                     condition_names = condition_names,
                                     genes = genes,
                                     batch_size = round(0.3*length(alpha_ind)),
                                     epochs = 1,
                                     learning_rate = learning_rate,
                                     regularization_rate = regularization_rate)
  }else{
    nn_alpha <- merge_many_nn_models(resfile = both_res_file[alpha_ind, ],
                                     mapfile = mapfile[alpha_ind, ],
                                     condition_names = condition_names,
                                     genes = genes,
                                     batch_size = round(0.3*length(alpha_ind)),
                                     epochs = epochs,
                                     learning_rate = learning_rate,
                                     regularization_rate = regularization_rate)
    
    
    sig_features_alpha <-
      nn_p_value_testing(
        nn_alpha,
        condition_name = condition_names,
        genotype_file = mapfile[alpha_ind, ],
        resistance_file = both_res_file[alpha_ind, ]
      )
    save(file='sig_features_alpha_16_drugs.RData',list = 'sig_features_alpha')
  }
   
  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)
  
  pruned_model_alpha <- nn_alpha
  set_weights(pruned_model_alpha, sig_features_alpha)
  
  
  pdf(file = 'nn_diagram_alpha.pdf',width = 6, height = 6)
  draw_model_diagram(name_model_weights(sig_features_alpha,genes,drugs),
                     vacuolar_transporters = c('YCF1','YBT1'),
                     membrane_transporters = c("PDR5", "YOR1", "SNQ2"),
                     angle_positions_membrane = c(90, 170,370) * (1/180) * pi,
                     angle_positions_vacuole = c(-20, 200) * (1/180) * pi)
  dev.off()
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  if('sig_features_A_16_drugs.RData' %in% list.files()){
    load('sig_features_A_16_drugs.RData')
    nn_A <- merge_many_nn_models(resfile = both_res_file[A_ind, ],
                                 mapfile = mapfile[A_ind, ],
                                 condition_names = condition_names,
                                 genes = genes,
                                 batch_size = round(0.3*length(A_ind)),
                                 epochs = 1,
                                 learning_rate = learning_rate,
                                 regularization_rate = regularization_rate)
  }else{
    nn_A <- merge_many_nn_models(resfile = both_res_file[A_ind, ],
                                 mapfile = mapfile[A_ind, ],
                                 condition_names = condition_names,
                                 genes = genes,
                                 batch_size = round(0.3*length(A_ind)),
                                 epochs = epochs,
                                 learning_rate = learning_rate,
                                 regularization_rate = regularization_rate)
    
    sig_features_A <-
      nn_p_value_testing(
        nn_A,
        condition_name = condition_names,
        genotype_file = mapfile[A_ind, ],
        resistance_file = both_res_file[A_ind, ]
      )
    save(file='sig_features_A_16_drugs.RData',list = 'sig_features_A')
  }
  
  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)
  
  pruned_model_A <- nn_A
  set_weights(pruned_model_A, sig_features_A)
  
  
  pdf(file = 'nn_diagram_A.pdf',width = 6, height = 6)
  draw_model_diagram(name_model_weights(sig_features_A,genes,drugs))
  dev.off()
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  if('sig_features_both_16_drugs.RData' %in% list.files()){
    load('sig_features_both_16_drugs.RData')
    nn_both <- merge_many_nn_models(resfile = both_res_file,
                                    mapfile = mapfile,
                                    condition_names = condition_names,
                                    genes = genes,
                                    batch_size = round(0.3*nrow(both_res_file)),
                                    epochs = 1,
                                    learning_rate = learning_rate,
                                    regularization_rate = regularization_rate)
  }else{
    nn_both <- merge_many_nn_models(resfile = both_res_file,
                                    mapfile = mapfile,
                                    condition_names = condition_names,
                                    genes = genes,
                                    batch_size = round(0.3*nrow(both_res_file)),
                                    epochs = epochs,
                                    learning_rate = learning_rate,
                                    regularization_rate = regularization_rate)
    sig_features_both <- nn_p_value_testing(
      nn_both,
      condition_name = condition_names,
      genotype_file = mapfile,
      resistance_file = both_res_file
    )
    save(file='sig_features_both_16_drugs.RData',list = 'sig_features_both')
  }

  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)

  pruned_model_both <- nn_both
  set_weights(pruned_model_both, sig_features_both)
  
  pdf(file = 'nn_diagram_both.pdf',width = 6, height = 6)
  #Cairo::CairoFonts(    
  #  regular="Arial:style=Regular",
  #  bold="Arial:style=Black",
  #  italic="Arial:style=Italic",
  #  bolditalic="Arial:style=Black Italic"#,
  #  #symbol="Symbol"
  #)
  draw_model_diagram(name_model_weights(sig_features_both,genes,drugs),
                     vacuolar_transporters = c('YCF1','YBT1','VMR1','BPT1','NFT1'),
                     membrane_transporters = c("PDR5", "YOR1", "SNQ2",'PDR18','YOL075C','PDR15','PDR12','PDR11','AUS1','PDR10','ADP1'),
                     angle_positions_vacuole = c(-20, 200, 245, 270, 295) * (1/180) * pi,
                     angle_positions_membrane = c(90, 170,370,200,220,240,260,280,300,320,340) * (1/180) * pi
                     )
  
  dev.off()
  
  named_weights <- name_model_weights(sig_features_both,genes,drugs)
  
  #Filter diagram for Fig 4
  named_weights_valinomycin <- named_weights[c(grep('YOR1',names(named_weights)),length(named_weights) -1)]
  named_weights_valinomycin$efflux_per_gene <- named_weights_valinomycin$efflux_per_gene[,'valinomycin',drop=F]
  
  
 
  
  pdf(file = 'nn_diagram_both_valinomycin_subset.pdf',width = 6, height = 6)
  draw_model_diagram(named_weights_valinomycin,
                     vacuolar_transporters = c('YBT1'),
                     membrane_transporters = c("YOR1", "PDR5", "SNQ2"),
                     angle_positions_vacuole = c(270) * (1/180) * pi,
                     angle_positions_membrane = c(90, 150, 190) * (1/180) * pi
  )
  dev.off()
  
  #Filter diagram for Fig 5
  named_weights_fluconazole <- named_weights[c(grep('PDR5',names(named_weights)),length(named_weights) -1)]
  named_weights_fluconazole$efflux_per_gene <- named_weights_fluconazole$efflux_per_gene[,'fluconazole',drop=F]
  
  pdf(file = 'nn_diagram_both_fluconazole_subset.pdf',width = 6, height = 6)
  draw_model_diagram(named_weights_fluconazole,
                     vacuolar_transporters = c('YCF1','YBT1'),
                     membrane_transporters = c("PDR5", "YOR1", "SNQ2"),
                     angle_positions_vacuole = c(-20, 200) * (1/180) * pi,
                     angle_positions_membrane = c(90, 190,350) * (1/180) * pi
  )
  dev.off()
  
  full_geno_list_A <- list()
  full_geno_list_alpha <- list()
  full_geno_list_both <- list()
  
  full_geno_list_A_extended <- list()
  full_geno_list_alpha_extended <- list()
  full_geno_list_both_extended <- list()
  for(i in 1:length(genes)){
    gene_name <- sprintf('%s_input',genes[i])
    full_geno_list_A[[gene_name]] <- 1 - mapfile[A_ind,genes[i]]
    full_geno_list_alpha[[gene_name]] <- 1 - mapfile[alpha_ind,genes[i]]
    full_geno_list_both[[gene_name]] <- 1 - mapfile[,genes[i]]
    
    full_geno_list_both_extended[[gene_name]] <- 1 - mapfile[,genes[i]]
  }
  full_geno_list_both_extended[['X_input']] <- rep(1, nrow(mapfile))
  
  
  
  
  
  Cairo::CairoPDF(file='weight_correlation.pdf',width=5,height=5)
  par(mar=c(5,5,1,1))
  plot((unlist(sig_features_A)),
       (unlist(sig_features_alpha)),
       pch = 16,
       col = rgb(0,0,0,1),
       xlab = 'Model weights (MATa)',
       ylab = parse(text='Model~weights~(MAT*alpha)'),
       cex.lab = 2,
       cex.axis = 2)
  abline(c(0,1),lty=2,lwd=1)
  correl <- cor((unlist(sig_features_A)),
      (unlist(sig_features_alpha)))
  correl <- format(correl,digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  text(-4,11,correl,cex=2,adj=0)
  dev.off()
  
  smoothscatter_legend <- function(){
    xm <- get('xm', envir = parent.frame(1))
    ym <- get('ym', envir = parent.frame(1))
    z  <- get('dens', envir = parent.frame(1))
    colramp <- get('colramp', parent.frame(1))
    fields::image.plot(xm,ym,z, col = colramp(256), legend.only = T, add =F, cex = 1.5)
  }
  
  
  png(file='mat_a_train.png',width=5.7,height=5,units='in',res=400,family='Arial')
  par(mar=c(5.5,5.5,1,5))
  smoothScatter(as.vector(predict(nn_A,full_geno_list_A)),
                as.vector(both_res_file[A_ind,]),
                xlab = 'Modelled Resistance (MATa)',
                ylab = 'Measured Resistance (MATa)',
                cex.lab = 1.5,
                cex.axis = 1.5,
                colramp = black_blue,
                postPlotHook = smoothscatter_legend)
  correl <- cor(as.vector(predict(nn_A,full_geno_list_A)),
                as.vector(both_res_file[A_ind,]))
  correl <- format(correl, digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  text(0,0.9,correl,cex=2,adj=0,col='white')
  abline(c(0,1),lwd=2,col='grey90',lty=5)
  dev.off()
  
  png(file='mat_alpha_train.png',width=5.7,height=5,units='in',res=400,family='Arial')
  par(mar=c(5.5,5.5,1,5))
  smoothScatter(as.vector(predict(nn_alpha,full_geno_list_alpha)),
                as.vector(both_res_file[alpha_ind,]),
                xlab = parse(text='Modelled~Resistance~(MAT*alpha)'),
                ylab = parse(text='Measured~Resistance~(MAT*alpha)'),
                cex.lab = 1.5,
                cex.axis = 1.5,
                colramp = black_blue,
                postPlotHook = smoothscatter_legend)
  correl <- cor(as.vector(predict(nn_alpha,full_geno_list_alpha)),
                as.vector(both_res_file[alpha_ind,]))
  correl <- format(correl, digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  text(0,0.9,correl,cex=2,adj=0,col='white')
  abline(c(0,1),lwd=2,col='grey90',lty=5)
  dev.off()
  
  
  png(file='mat_a_train_mat_alpha_test.png',width=5.7,height=5,units='in',res=400,family='Arial')
  par(mar=c(5.5,5.5,1,5))
  smoothScatter(as.vector(predict(nn_A,full_geno_list_alpha)),
                as.vector(both_res_file[alpha_ind,]),
                xlab = parse(text='Modelled~Resistance~(MAT*a)'),
                ylab = parse(text='Measured~Resistance~(MAT*alpha)'),
                cex.lab = 1.5,
                cex.axis = 1.5,
                colramp = black_blue,
                postPlotHook = smoothscatter_legend)
  correl <- cor(as.vector(predict(nn_A,full_geno_list_alpha)),
                as.vector(both_res_file[alpha_ind,]))
  correl <- format(correl, digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  text(0,0.9,correl,cex=2,adj=0,col='white')
  abline(c(0,1),lwd=2,col='grey90',lty=5)
  dev.off()
  
  
  png(file='mat_alpha_train_mat_a_test.png',width=5.7,height=5,units='in',res=400,family='Arial')
  par(mar=c(5.5,5.5,1,5))
  smoothScatter(as.vector(predict(nn_alpha,full_geno_list_A)),
                as.vector(both_res_file[A_ind,]),
                xlab = parse(text='Modelled~Resistance~(MAT*alpha)'),
                ylab = parse(text='Measured~Resistance~(MAT*a)'),
                cex.lab = 1.5,
                cex.axis = 1.5,
                colramp = black_blue,
                postPlotHook = smoothscatter_legend)
  correl <- cor(as.vector(predict(nn_alpha,full_geno_list_A)),
                as.vector(both_res_file[A_ind,]))
  correl <- format(correl, digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  text(0,0.9,correl,cex=2,adj=0,col='white')
  abline(c(0,1),lwd=2,col='grey90',lty=5)
  dev.off()
  
  png(file='both_train.png',width=5.7,height=5,units='in',res=400,family='Arial')
  par(mar=c(5.5,5.5,1,5))
  smoothScatter(as.vector(predict(nn_both,full_geno_list_both)),
                as.vector(both_res_file),
                xlab = parse(text='Modelled~Resistance'),
                ylab = parse(text='Measured~Resistance'),
                cex.lab = 1.8,
                cex.axis = 1.5,
                colramp = black_blue,
                postPlotHook = smoothscatter_legend)
  correl <- cor(as.vector(predict(nn_both,full_geno_list_both)),
                as.vector(both_res_file))
  correl <- format(correl, digits = 2)
  correl <- paste(c('r = ',correl),collapse='')
  #text(0.1,0.8,correl,cex=2,adj=0,col='white')
  abline(c(0,1),lwd=2,col='grey90',lty=5)
  dev.off()
  
  
  dir.create('single_drug_performance', showWarnings = FALSE)
  setwd('single_drug_performance')
  
  for(drug in drugs) {
    filename <- sprintf('nn_performance_%s_both.pdf',drug)
    pdf(file = filename,width = 4, height = 4)
    compare_nn_predictions(
      nn_both,
      mapfile,
      both_res_file,
      genes_to_predict = genes,
      drug = drug,
      gene_palette = my_gene_colors
    )
    abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
    dev.off()
    
    #Make a zoomed-in plot for fluconazole and valinomycin
    if(drug %in% c('valinomycin','fluconazole')){
      filename <- sprintf('nn_performance_%s_both_autoscale.pdf',drug)
      pdf(file = filename,width = 4, height = 4)
      compare_nn_predictions(
        nn_both,
        mapfile,
        both_res_file,
        genes_to_predict = genes,
        drug = drug,
        gene_palette = my_gene_colors,
        xlims = NULL,
        ylims = NULL
        
      )
      abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
      dev.off()
    }
    
  }
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  save('named_weights',file='named_weights_for_nn')
  write(format_named_weights_to_string(named_weights),file='named_weights_string_for_nn.txt')
  
}

if('NN extensions' %in% to_run | 'all' %in% to_run){
  set.seed(123)
  setwd(this.dir)
  
  
  setwd(input_data_directory)
  mapping_file <- read.table(mapping_filename, head = T, row.names = 1)
  
  setwd(this.dir)
  setwd(output_data_directory)
  both_res_filename <- paste(c(resistance_output_prefix,'_both.tsv'),collapse='')
  both_res_file <- read.table(both_res_filename)
  
  #Scale to 0-1 interval
  both_res_file[both_res_file < 1e-10] <- 1e-10
  both_res_file <- apply(both_res_file,2,function(x){x/max(x)})
  
  mapfile <- mapping_file[rownames(both_res_file),]
  
  genes <- colnames(mapfile)[2:17]
  condition_names <- colnames(both_res_file)
  drugs <- sapply(colnames(both_res_file),function(x){strsplit(x,split='_')[[1]][1]})
  
  alpha_ind <- grep('alpha',mapfile$Plate)
  A_ind <- grep('alpha',mapfile$Plate, invert = T)
  
  nn_output_path <-
    paste(c(output_results_directory,nn_output_dir,'extensions'), collapse = '/')
  setwd(this.dir)
  setwd(input_data_directory)
  dir.create(nn_output_path, showWarnings = FALSE)
  setwd(nn_output_path)
  
  ##Make a NN with a hidden efflux factor for valinomycin
  extended_mapfile <- cbind(mapfile,rep(0,nrow(mapfile)))
  colnames(extended_mapfile)[ncol(extended_mapfile)] <- 'X'
  
  ##Make with extra neuron
  learning_rate <- 0.01
  epochs <- 10000
  regularization_rate <- 5e-04
  
  for(drug in c('valinomycin')){

    
    setwd(this.dir)
    setwd(output_data_directory)
    filename <- sprintf('sig_features_both_%s_extended.RData',drug)
    
    if(filename %in% list.files()){
      nn_both_extended <- merge_many_nn_models(
        resfile = both_res_file,
        mapfile = extended_mapfile,
        condition_names = grep(drug, condition_names, val = T),
        genes = c(genes,'X'),
        batch_size = round(0.3*nrow(both_res_file)),
        epochs = 1,
        learning_rate = learning_rate,
        regularization_rate = regularization_rate
        
        
      )
      load(filename)
              
      
    }else{
      nn_both_extended <- merge_many_nn_models(
        resfile = both_res_file,
        mapfile = extended_mapfile,
        condition_names = grep(drug, condition_names, val = T),
        genes = c(genes,'X'),
        batch_size = round(0.3*nrow(both_res_file)),
        epochs = epochs,
        learning_rate = learning_rate,
        regularization_rate = regularization_rate
      )
      sig_features_both_extended <- nn_p_value_testing(
        nn_both_extended,
        condition_name = grep(drug, condition_names, val = T),
        genotype_file = extended_mapfile,
        resistance_file = both_res_file,
        genes = c(genes, 'X')
      )
      
      save(file=filename,list = 'sig_features_both_extended')
      
    }
    weight_string <- format_named_weights_to_string(name_model_weights(sig_features_both_extended, c(genes, 'X'), drug))
    write(weight_string,file=sprintf('named_weights_string_for_nn_%s_extended.txt',drug))
   
    pruned_model_both_extended <- nn_both_extended
    set_weights(pruned_model_both_extended, sig_features_both_extended)
    
    setwd(this.dir)
    setwd(input_data_directory)
    setwd(nn_output_path)
    
    pdf(file = sprintf('nn_diagram_%s_both_extended.pdf',drug),
        width = 6,
        height = 6)
    draw_model_diagram(
      name_model_weights(sig_features_both_extended, c(genes, 'X'), drug),
      vacuolar_transporters = c('YBT1','YCF1'),
      membrane_transporters = c("YOR1", "PDR5", "SNQ2", 'X'),
      big_transporters = c("PDR5", "YOR1", "SNQ2", "YBT1", 'YCF1',"X"),
      angle_positions_membrane = c(90, 150, 190, 370) * (1 /
                                                           180) * pi,
      angle_positions_vacuole = c(220,320) * (1 / 180) * pi,
      transporter_colors = list(
        PDR5 = rgb(41, 145, 206, maxColorValue = 255),
        SNQ2 = rgb(180, 231, 172, maxColorValue = 255),
        YOR1 = rgb(234, 151, 73, maxColorValue = 255),
        YBT1 = rgb(255, 255, 190, maxColorValue = 255),
        YCF1 = rgb(167, 0, 6, maxColorValue = 255),
        X = rgb(255, 255, 255, maxColorValue = 255)
      )
    )
    dev.off()
    
    pdf(file = sprintf('nn_performance_%s_both_extended.pdf',drug),
        width = 4,
        height = 4)
    compare_nn_predictions(
      pruned_model_both_extended,
      mapfile,
      both_res_file[, sprintf('%s_both',drug), drop = F],
      genes_to_predict = c(genes, 'X'),
      drug = drug,
      gene_palette = my_gene_colors,
      xlim = NULL,
      ylim = NULL
    )
    abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
    dev.off()
    
    ##Compare to NN without extra
    
    setwd(this.dir)
    setwd(output_data_directory)
    filename <- sprintf('sig_features_both_%s_single_environment.RData',drug)
    
    if(filename %in% list.files()){
      nn_both <- merge_many_nn_models(
        resfile = both_res_file,
        mapfile = extended_mapfile,
        condition_names = grep(drug, condition_names, val = T),
        genes = genes,
        batch_size = round(0.3*nrow(both_res_file)),
        epochs = 1,
        learning_rate = learning_rate,
        regularization_rate = regularization_rate
      )
      load(filename)
    }else{
      nn_both <- merge_many_nn_models(
        resfile = both_res_file,
        mapfile = extended_mapfile,
        condition_names = grep(drug, condition_names, val = T),
        genes = genes,
        batch_size = round(0.3*nrow(both_res_file)),
        epochs = epochs,
        learning_rate = learning_rate,
        regularization_rate = regularization_rate
      )
      sig_features_both <- nn_p_value_testing(
        nn_both,
        condition_name = grep(drug, condition_names, val = T),
        genotype_file = extended_mapfile,
        resistance_file = both_res_file,
        genes = genes
      )
      save(file=filename,list = 'sig_features_both')
    }
    
    weight_string <- format_named_weights_to_string(name_model_weights(sig_features_both, genes, drug))
    write(weight_string,file=sprintf('named_weights_string_for_nn_%s_single_environment.txt',drug))
    
      
    setwd(this.dir)
    setwd(input_data_directory)
    setwd(nn_output_path)
    
    
    pruned_model_both <- nn_both
    set_weights(pruned_model_both, sig_features_both)
    
    pdf(file = sprintf('nn_performance_%s_both_single_environment_training.pdf',drug),
        width = 4,
        height = 4)
    compare_nn_predictions(
      pruned_model_both,
      mapfile,
      both_res_file[,sprintf('%s_both',drug),drop=F],
      genes_to_predict = genes,
      drug = drug,
      genes_to_split = c('SNQ2', 'PDR5', 'YBT1', 'YCF1', 'YOR1'),
      gene_palette = my_gene_colors,
      xlim = NULL,
      ylim = NULL
    )
    abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
    dev.off()
    
    pdf(file = sprintf('nn_diagram_%s_both_single_environment_training.pdf',drug),
        width = 6,
        height = 6)
    draw_model_diagram(
      name_model_weights(sig_features_both, genes, drug),
      vacuolar_transporters = c('YCF1','YBT1'),
      membrane_transporters = c("PDR5", "YOR1", "SNQ2"),
      angle_positions_membrane = c(90, 170,370) * (1/180) * pi,
      angle_positions_vacuole = c(-20, 200) * (1/180) * pi
    )
    dev.off()
  }
  
  
  
  #Make extension for fluconazole
  modifier_genes <- c('PDR5', 'SNQ2', 'YBT1', 'YCF1', 'YOR1')
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  filename <- 'fluc_three_layer_model_reg_search.RData'
  
  if(filename %in% list.files()){
    load(filename)
  }else{
    search_reg_rate_three_layer <- nn_model_search(resfile = both_res_file,
                                                   mapfile = mapfile,
                                                   genes = modifier_genes,
                                                   efflux_genes = 'PDR5',
                                                   condition_names = 'fluconazole_both',
                                                   batch_size = round(0.3*nrow(both_res_file)),
                                                   weight_merging_function = average_weight_list,
                                                   nn_function = make_three_layer_nn_model,
                                                   epochs = 10000,
                                                   learning_rate = 0.01,
                                                   regularization_rates = 10^seq(-6,0,by=0.5))
    save(file=filename, list = 'search_reg_rate_three_layer')
  }
  

  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)
  
  
  pdf(file = 'mse_vs_reg_rate_three_layer_fluc.pdf',
      width = 4.5,
      height = 3.5)
  par(mar = c(4.5, 4.5, 1, 1))
  plot(
    log10(search_reg_rate_three_layer[, 1]),
    (search_reg_rate_three_layer[, 2]),
    type = 'l',
    col = 'purple',
    xlab = expression(paste("log" ["10"], " (Regulariation rate)")),
    ylab = 'Mean Squared Error',
    cex.lab = 1.5,
    lwd = 2
  )
  abline(v = -5, lty = 2)
  dev.off()
  
  
  
  setwd(this.dir)
  setwd(output_data_directory)
  filename <- 'fluc_three_layer_model.RData'
  
  if(filename %in% list.files()){
    nn_fluc_three_layer <- merge_many_nn_models(resfile = both_res_file,
                                                mapfile = mapfile,
                                                genes = modifier_genes,
                                                efflux_genes = 'PDR5',
                                                condition_names = 'fluconazole_both',
                                                batch_size = round(0.3*nrow(both_res_file)),
                                                weight_merging_function = average_weight_list,
                                                nn_function = make_three_layer_nn_model,
                                                epochs = 1,
                                                learning_rate = 0.01,
                                                regularization_rate = 1e-05
    )
    load(filename)
  }else{
    nn_fluc_three_layer <- merge_many_nn_models(resfile = both_res_file,
                                                mapfile = mapfile,
                                                genes = modifier_genes,
                                                efflux_genes = 'PDR5',
                                                condition_names = 'fluconazole_both',
                                                batch_size = round(0.3*nrow(both_res_file)),
                                                weight_merging_function = average_weight_list,
                                                nn_function = make_three_layer_nn_model,
                                                epochs = 10000,
                                                learning_rate = 0.01,
                                                regularization_rate = 1e-05
    )
    
    sig_features_three_layer <- nn_p_value_testing(
      nn_fluc_three_layer,
      condition_name = 'fluconazole_both',
      genotype_file = mapfile,
      resistance_file = both_res_file,
      genes = modifier_genes,
      efflux_genes = 'PDR5',
      nn_model_function = make_three_layer_nn_model
    )
    
    save(file = filename, list = 'sig_features_three_layer')
  }
  
  
  set_weights(nn_fluc_three_layer, sig_features_three_layer)
  named_three_layer_weights <-
    name_model_weights(
      get_weights(nn_fluc_three_layer),
      modifier_genes,
      'fluconazole',
      three_layers = T,
      efflux_genes = c('PDR5')
    )
  
  
  write(format_named_weights_to_string(named_three_layer_weights),file='named_weights_string_for_nn_fluconazole_three_layer.txt')

    
  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)
  
  
 
  
  pdf(file = sprintf('nn_performance_fluconazole_three_layer_both_single_environment_training.pdf',drug),
      width = 4,
      height = 4)
  compare_nn_predictions(
    nn_fluc_three_layer,
    mapfile,
    both_res_file[,'fluconazole_both',drop=F],
    genes_to_predict = c('PDR5','SNQ2','YBT1','YCF1','YOR1'),
    drug = 'fluconazole',
    genes_to_split = c('SNQ2', 'PDR5', 'YBT1', 'YCF1', 'YOR1'),
    gene_palette = my_gene_colors,
    xlim = NULL,
    ylim = NULL
  ) 
  abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
  dev.off()
  
  
  named_three_layer_weights <-
    name_model_weights(
      get_weights(nn_fluc_three_layer),
      modifier_genes,
      'fluconazole',
      three_layers = T,
      efflux_genes = c('PDR5')
    )
  
  pdf(file = 'nn_diagram_fluconazole_indirect_original.pdf',width = 6, height = 6)
  draw_model_diagram(
    named_three_layer_weights,
    membrane_transporters = c("PDR5", "YOR1",
                              "PDR5_indirect", "SNQ2"),
    vacuolar_transporters = c("YCF1", "YBT1"),
    angle_positions_vacuole = c(-20, 200) * (1/180) * pi,
    angle_positions_membrane = c(90, 190, 45, 350) * (1/180) * pi,
    transporter_radius_small = 0.15,
    influence_line_width_max_size = 10,
    #inhib
    transporter_colors = list('PDR5' = rgb(41,145, 206, maxColorValue = 255),
                              'SNQ2' = rgb(180, 231, 172,  maxColorValue = 255),
                              'YOR1' = rgb(234, 151, 73, maxColorValue = 255), 
                              'YBT1' = rgb(255, 255, 190, maxColorValue = 255),
                              'YCF1' = rgb(167, 0, 6, maxColorValue = 255),
                              'PDR5_indirect' = rgb(0, 0, 0, maxColorValue = 255)),
    i_weights_in_legend = c(-0.04,-0.08,-0.16,2.4),
    legend_size_left = 0.4,
    influence_line_width_scale_factor = 20,
    influence_arrowhead_scale_factor = 0.05
    
  )
  dev.off()
  
  pdf(file = 'nn_diagram_fluconazole_indirect_gradient_scaled.pdf',width = 6, height = 6)
  draw_model_diagram(
    named_three_layer_weights,
    membrane_transporters = c("PDR5", "YOR1",
                              "PDR5_indirect", "SNQ2"),
    vacuolar_transporters = c("YCF1", "YBT1"),
    angle_positions_vacuole = c(-20, 200) * (1/180) * pi,
    angle_positions_membrane = c(90, 190, 45, 350) * (1/180) * pi,
    transporter_radius_small = 0.15,
    influence_line_width_max_size = Inf,
    #inhib
    transporter_colors = list('PDR5' = rgb(41,145, 206, maxColorValue = 255),
                              'SNQ2' = rgb(180, 231, 172,  maxColorValue = 255),
                              'YOR1' = rgb(234, 151, 73, maxColorValue = 255), 
                              'YBT1' = rgb(255, 255, 190, maxColorValue = 255),
                              'YCF1' = rgb(167, 0, 6, maxColorValue = 255),
                              'PDR5_indirect' = rgb(0, 0, 0, maxColorValue = 255)),
    i_weights_in_legend = c(-0.16,-0.64,-2.56),
    legend_size_left = 0.3,
    influence_line_width_scale_factor = 5*get_relative_third_layer_activities(nn_fluc_three_layer),
    influence_arrowhead_scale_factor = 0.05
    
  )
  dev.off()
  
  ##Compare to baseline
  setwd(this.dir)
  setwd(output_data_directory)
  filename <- 'fluc_single_environment_model.RData'
  
  
  if(filename %in% list.files()){
    arglist <-
      list(
        'resistance_file' = both_res_file,
        'genotype_file' = mapfile,
        'condition_name' = 'fluconazole_both',
        'genes' = genes,
        'efflux_genes' = genes,
        'regularization_rate' = 5e-04,
        'learning_rate' = 0.05,
        'epochs' = 1,
        'batch_size' = round(0.3*nrow(mapfile))
      )
    
    use_session_with_seed(1234)
    weight_list <- list()
    for (i in 1:2) {
      nn_fluc <- do.call(make_nn_model, arglist)
      weight_list[[length(weight_list) + 1]] <-
        get_weights(nn_fluc$model)
    }
    averaged_weight_list <- prune_weight_list(weight_list)
    set_weights(nn_fluc$model, averaged_weight_list)
    
    load(filename)
    
  }else{
    arglist <-
      list(
        'resistance_file' = both_res_file,
        'genotype_file' = mapfile,
        'condition_name' = 'fluconazole_both',
        'genes' = genes,
        'efflux_genes' = genes,
        'regularization_rate' = 5e-04,
        'learning_rate' = 0.05,
        'epochs' = 10000,
        'batch_size' = round(0.3*nrow(mapfile))
      )
    
    use_session_with_seed(1234)
    weight_list <- list()
    for (i in 1:10) {
      nn_fluc <- do.call(make_nn_model, arglist)
      weight_list[[length(weight_list) + 1]] <-
        get_weights(nn_fluc$model)
    }
    averaged_weight_list <- prune_weight_list(weight_list)
    set_weights(nn_fluc$model, averaged_weight_list)
    
    sig_features <- nn_p_value_testing(
      nn_fluc$model,
      genotype_file = mapfile,
      resistance_file = both_res_file,
      condition_name = 'fluconazole_both',
      nn_model_function = make_nn_model,
      nn_model_parameters = arglist
    )
    
    save(file = filename, list = 'sig_features')
  }
  
  set_weights(nn_fluc$model,sig_features)
  named_weights_fluc_single_envir <- name_model_weights(get_weights(nn_fluc$model), genes, 'fluconazole')
  write(format_named_weights_to_string(named_weights_fluc_single_envir),file='named_weights_string_for_nn_fluconazole_single_environment.txt')
  
  
  setwd(this.dir)
  setwd(input_data_directory)
  setwd(nn_output_path)
  
  
  pdf(file = sprintf('nn_performance_fluconazole_both_single_environment_training.pdf',drug),
      width = 4,
      height = 4)
  compare_nn_predictions(
    nn_fluc$model,
    mapfile,
    both_res_file[,'fluconazole_both',drop=F],
    genes_to_predict = genes,
    drug = 'fluconazole',
    genes_to_split = c('SNQ2', 'PDR5', 'YBT1', 'YCF1', 'YOR1'),
    gene_palette = my_gene_colors,
    xlim = NULL,
    ylim = NULL
  )
  abline(c(0,1),lty=5,col=rgb(0,0,0,0.5))
  dev.off()
  
  
}

if ('qRT-PCR' %in% to_run | 'all' %in% to_run){
  setwd(this.dir)
  setwd(input_data_directory)
  rt_data <- xlsx::read.xlsx('RT_PCR_Dec52017.xlsx',sheetIndex = 1)
  
  qrt_pcr_output_path <-
    paste(c(output_results_directory,qrt_pcr_output_path), collapse = '/')
  
  dir.create(qrt_pcr_output_path, showWarnings = FALSE)
  
  setwd(qrt_pcr_output_path)
  
  if('NN extensions' %in% to_run){
    nn_overlay = T
  }else{
    nn_overlay = F
  }
  
  Cairo::CairoPDF(file='qrt_pcr_results',width = 8,height = 7)
  Cairo::CairoFonts(    
    regular="Arial:style=Regular",
    bold="Arial:style=Black",
    italic="Arial:style=Italic",
    bolditalic="Arial:style=Black Italic",
    symbol="Symbol"
  )
  make_rt_pcr_plot(rt_data,
                   overlay_nn_results = nn_overlay,
                   nn_results = named_three_layer_weights)
  
  
  dev.off()
}
